<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Anna Korda">

<title>Hands on I - Supervised Learning</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Hands_on_I_files/libs/clipboard/clipboard.min.js"></script>
<script src="Hands_on_I_files/libs/quarto-html/quarto.js"></script>
<script src="Hands_on_I_files/libs/quarto-html/popper.min.js"></script>
<script src="Hands_on_I_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Hands_on_I_files/libs/quarto-html/anchor.min.js"></script>
<link href="Hands_on_I_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Hands_on_I_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Hands_on_I_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Hands_on_I_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Hands_on_I_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


<link rel="stylesheet" href="styles.css">
</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#analysis-of-the-heart-disease-dataset" id="toc-analysis-of-the-heart-disease-dataset" class="nav-link active" data-scroll-target="#analysis-of-the-heart-disease-dataset">1. Analysis of the Heart Disease Dataset</a>
  <ul class="collapse">
  <li><a href="#perform-an-eda-on-the-dataset" id="toc-perform-an-eda-on-the-dataset" class="nav-link" data-scroll-target="#perform-an-eda-on-the-dataset">1.1. Perform an EDA on the dataset</a>
  <ul class="collapse">
  <li><a href="#data-cleaning-and-missing-data" id="toc-data-cleaning-and-missing-data" class="nav-link" data-scroll-target="#data-cleaning-and-missing-data">1.1.1. Data cleaning and missing data</a></li>
  </ul></li>
  <li><a href="#create-visualizations-in-order-to-show-which-variables-seem-to-be-more-associated-with-heart-disease" id="toc-create-visualizations-in-order-to-show-which-variables-seem-to-be-more-associated-with-heart-disease" class="nav-link" data-scroll-target="#create-visualizations-in-order-to-show-which-variables-seem-to-be-more-associated-with-heart-disease">1.2. Create visualizations in order to show which variables seem to be more associated with heart disease</a></li>
  </ul></li>
  <li><a href="#difference-in-mortality-rates-in-hospitalized-covid-19-patients" id="toc-difference-in-mortality-rates-in-hospitalized-covid-19-patients" class="nav-link" data-scroll-target="#difference-in-mortality-rates-in-hospitalized-covid-19-patients">2. Difference in mortality rates in hospitalized COVID-19 patients</a>
  <ul class="collapse">
  <li><a href="#reproduce-figure-1-from-the-publication" id="toc-reproduce-figure-1-from-the-publication" class="nav-link" data-scroll-target="#reproduce-figure-1-from-the-publication">2.1. Reproduce Figure 1 from the publication</a></li>
  <li><a href="#reproduce-figure-2-from-the-publication-but-instead-of-representing-the-clusters-in-the-annotation-represent-the-groups-g1-to-g4" id="toc-reproduce-figure-2-from-the-publication-but-instead-of-representing-the-clusters-in-the-annotation-represent-the-groups-g1-to-g4" class="nav-link" data-scroll-target="#reproduce-figure-2-from-the-publication-but-instead-of-representing-the-clusters-in-the-annotation-represent-the-groups-g1-to-g4">2.2. Reproduce Figure 2 from the publication but instead of representing the clusters in the annotation, represent the groups (G1 to G4)</a>
  <ul class="collapse">
  <li><a href="#loading-and-cleaning-the-data" id="toc-loading-and-cleaning-the-data" class="nav-link" data-scroll-target="#loading-and-cleaning-the-data">2.2.1. Loading and cleaning the data</a></li>
  <li><a href="#heat-map" id="toc-heat-map" class="nav-link" data-scroll-target="#heat-map">2.2.3. Heat map</a></li>
  </ul></li>
  <li><a href="#improve-figure-2-of-the-publication" id="toc-improve-figure-2-of-the-publication" class="nav-link" data-scroll-target="#improve-figure-2-of-the-publication">2.3. Improve Figure 2 of the publication</a></li>
  </ul></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">session info</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="Hands_on_I.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Hands on I - Supervised Learning</h1>
<p class="subtitle lead">Data Mining and Data Integration in Biomedicine</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Anna Korda </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">March, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<div style="page-break-after: always;"></div>
<section id="analysis-of-the-heart-disease-dataset" class="level1">
<h1>1. Analysis of the Heart Disease Dataset</h1>
<p>Load the data from <a href="https://raw.githubusercontent.com/jpinero/DMI_2021/main/datasets/heart_disease_dataset.csv">here</a>, and the description is <a href="https://raw.githubusercontent.com/jpinero/DMI_2021/main/datasets/heart_disease_description.txt">here</a>. The original dataset comes from <a href="https://archive.ics.uci.edu/ml/datasets/Heart+Disease">here</a> and corresponds to the <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data">processed cleveland data</a></p>
<section id="perform-an-eda-on-the-dataset" class="level2">
<h2 class="anchored" data-anchor-id="perform-an-eda-on-the-dataset">1.1. Perform an EDA on the dataset</h2>
<p>We will first start by reading the file containing our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>dataset <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="at">file =</span> <span class="st">"data/heart_disease_dataset.csv"</span>, <span class="at">sep =</span> <span class="st">" "</span>, <span class="at">header =</span> T, <span class="at">dec =</span> <span class="st">"."</span>, <span class="at">stringsAsFactors =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see our data has 303 rows, corresponding to individuals or patients, and 15 columns, one that identifies each individual and 14 different attributes.</p>
<p>The description of each attribute can be checked here:</p>
<ul>
<li><p><strong>Age</strong>: age in years.</p></li>
<li><p><strong>Sex</strong>: sex (1 = male; 0 = female).</p></li>
<li><p><strong>Cp</strong>: chest pain type.</p>
<ul>
<li>Value 1: typical angina.</li>
<li>Value 2: atypical angina.</li>
<li>Value 3: non-anginal pain.</li>
<li>Value 4: asymptomatic.</li>
</ul></li>
<li><p><strong>Trestbps</strong>: resting blood pressure (in mm Hg on admission to the hospital).</p></li>
<li><p><strong>Chol</strong>: serum cholestoral in mg/dl.</p></li>
<li><p><strong>Fbs</strong>: fasting blood sugar &gt; 120 mg/dl (1 = true; 0 = false)</p></li>
<li><p><strong>Restecg</strong>: resting electrocardiographic results.</p>
<ul>
<li>Value 0: normal.</li>
<li>Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of &gt; 0.05 mV).</li>
<li>Value 2: showing probable or definite left ventricular hypertrophy by Estes’ criteria.</li>
</ul></li>
<li><p><strong>Thalach</strong>: maximum heart rate achieved.</p></li>
<li><p><strong>Exang</strong>: exercise induced angina (1 = yes; 0 = no).</p></li>
<li><p><strong>Oldpeak</strong>: T depression induced by exercise relative to rest.</p></li>
<li><p><strong>Slope</strong>: the slope of the peak exercise ST segment:</p>
<ul>
<li>Value 1: upsloping.</li>
<li>Value 2: flat.</li>
<li>Value 3: downsloping.</li>
</ul></li>
<li><p><strong>Ca</strong>: number of major vessels (0-3) colored by flourosopy.</p></li>
<li><p><strong>Thal</strong>: 3 = normal; 6 = fixed defect; 7 = reversable defect.</p></li>
<li><p><strong>Num</strong>: diagnosis of heart disease (angiographic disease status).</p>
<ul>
<li>Value 0: &lt; 50% diameter narrowing.</li>
<li>Value 1: &gt; 50% diameter narrowing.</li>
</ul></li>
</ul>
<section id="data-cleaning-and-missing-data" class="level3">
<h3 class="anchored" data-anchor-id="data-cleaning-and-missing-data">1.1.1. Data cleaning and missing data</h3>
<p>Once we have our data and we have familiarized ourselves with its attributes, we should check for any potential missing/incorrect data. To do so we can use the package <code>dplyr</code> and <code>fauxanif</code>, which includes functions such as <code>na_if_in()</code> or <code>na_if_not()</code>. We will “transform” any missing data, represented with a -9, as well as any data that we consider incorrect or improbable, with an <code>NA</code> value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="fu">library</span>(fauxnaif)</span>
<span id="cb2-3"><a href="#cb2-3"></a></span>
<span id="cb2-4"><a href="#cb2-4"></a>dataset <span class="ot">&lt;-</span> dataset <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(</span>
<span id="cb2-5"><a href="#cb2-5"></a>  <span class="at">age =</span> <span class="fu">na_if_not</span>(age, <span class="dv">18</span><span class="sc">:</span><span class="dv">100</span>), </span>
<span id="cb2-6"><a href="#cb2-6"></a>  <span class="fu">across</span>(<span class="fu">c</span>(sex, fbs, exang), na_if_not, <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)),</span>
<span id="cb2-7"><a href="#cb2-7"></a>  <span class="at">cp =</span> <span class="fu">na_if_not</span>(cp, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>),</span>
<span id="cb2-8"><a href="#cb2-8"></a>  <span class="at">trestbps =</span> <span class="fu">na_if_not</span>(trestbps, <span class="dv">80</span><span class="sc">:</span><span class="dv">200</span>),</span>
<span id="cb2-9"><a href="#cb2-9"></a>  <span class="co"># ? chol = na_if_not()</span></span>
<span id="cb2-10"><a href="#cb2-10"></a>  <span class="at">restecg =</span> <span class="fu">na_if_not</span>(restecg, <span class="dv">0</span><span class="sc">:</span><span class="dv">2</span>),</span>
<span id="cb2-11"><a href="#cb2-11"></a>  <span class="at">thalach =</span> <span class="fu">na_if_not</span>(thalach, <span class="dv">60</span><span class="sc">:</span><span class="dv">210</span>),</span>
<span id="cb2-12"><a href="#cb2-12"></a>  <span class="at">oldpeak =</span> <span class="fu">na_if_in</span>(oldpeak, <span class="sc">~</span> . <span class="sc">&lt;</span> <span class="dv">0</span>),</span>
<span id="cb2-13"><a href="#cb2-13"></a>  <span class="at">slope =</span> <span class="fu">na_if_not</span>(slope, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>),</span>
<span id="cb2-14"><a href="#cb2-14"></a>  <span class="at">ca =</span> <span class="fu">na_if_not</span>(ca, <span class="dv">0</span><span class="sc">:</span><span class="dv">3</span>), </span>
<span id="cb2-15"><a href="#cb2-15"></a>  <span class="at">thal =</span> <span class="fu">na_if_not</span>(thal, <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">6</span>,<span class="dv">7</span>)),</span>
<span id="cb2-16"><a href="#cb2-16"></a>  <span class="at">num =</span> <span class="fu">na_if_not</span>(num, <span class="dv">0</span><span class="sc">:</span><span class="dv">4</span>)</span>
<span id="cb2-17"><a href="#cb2-17"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Afterwards, we will use the package <code>mice</code> to replace the missing values an estimated value. Before that, we can observe the missing data pattern using the function <code>md.pattern()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">library</span>(mice)</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="fu">md.pattern</span>(dataset, <span class="at">rotate.names =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    age sex cp trestbps chol fbs restecg thalach exang oldpeak slope num
297   1   1  1        1    1   1       1       1     1       1     1   1
4     1   1  1        1    1   1       1       1     1       1     1   1
2     1   1  1        1    1   1       1       1     1       1     1   1
      0   0  0        0    0   0       0       0     0       0     0   0
    patient_id thal ca  
297          1    1  1 0
4            1    1  0 1
2            1    0  1 1
             0    2  4 6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>dataset<span class="sc">$</span>ca <span class="ot">&lt;-</span> <span class="fu">factor</span>(dataset<span class="sc">$</span>ca, <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>))</span>
<span id="cb5-2"><a href="#cb5-2"></a>dataset<span class="sc">$</span>thal <span class="ot">&lt;-</span> <span class="fu">factor</span>(dataset<span class="sc">$</span>thal, <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">6</span>, <span class="dv">7</span>))</span>
<span id="cb5-3"><a href="#cb5-3"></a>methods <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">"pmm"</span>,<span class="st">"polyreg"</span>,<span class="st">""</span>,<span class="st">""</span>)</span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="fu">names</span>(methods) <span class="ot">&lt;-</span> <span class="fu">names</span>(dataset)</span>
<span id="cb5-5"><a href="#cb5-5"></a>imputed_data <span class="ot">&lt;-</span> <span class="fu">mice</span>(dataset, <span class="at">method =</span> methods, <span class="at">m =</span> <span class="dv">5</span>, <span class="at">maxit =</span> <span class="dv">100</span>, <span class="at">seed =</span> <span class="dv">123</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 iter imp variable
  1   1  ca  thal
  1   2  ca  thal
  1   3  ca  thal
  1   4  ca  thal
  1   5  ca  thal
  2   1  ca  thal
  2   2  ca  thal
  2   3  ca  thal
  2   4  ca  thal
  2   5  ca  thal
  3   1  ca  thal
  3   2  ca  thal
  3   3  ca  thal
  3   4  ca  thal
  3   5  ca  thal
  4   1  ca  thal
  4   2  ca  thal
  4   3  ca  thal
  4   4  ca  thal
  4   5  ca  thal
  5   1  ca  thal
  5   2  ca  thal
  5   3  ca  thal
  5   4  ca  thal
  5   5  ca  thal
  6   1  ca  thal
  6   2  ca  thal
  6   3  ca  thal
  6   4  ca  thal
  6   5  ca  thal
  7   1  ca  thal
  7   2  ca  thal
  7   3  ca  thal
  7   4  ca  thal
  7   5  ca  thal
  8   1  ca  thal
  8   2  ca  thal
  8   3  ca  thal
  8   4  ca  thal
  8   5  ca  thal
  9   1  ca  thal
  9   2  ca  thal
  9   3  ca  thal
  9   4  ca  thal
  9   5  ca  thal
  10   1  ca  thal
  10   2  ca  thal
  10   3  ca  thal
  10   4  ca  thal
  10   5  ca  thal
  11   1  ca  thal
  11   2  ca  thal
  11   3  ca  thal
  11   4  ca  thal
  11   5  ca  thal
  12   1  ca  thal
  12   2  ca  thal
  12   3  ca  thal
  12   4  ca  thal
  12   5  ca  thal
  13   1  ca  thal
  13   2  ca  thal
  13   3  ca  thal
  13   4  ca  thal
  13   5  ca  thal
  14   1  ca  thal
  14   2  ca  thal
  14   3  ca  thal
  14   4  ca  thal
  14   5  ca  thal
  15   1  ca  thal
  15   2  ca  thal
  15   3  ca  thal
  15   4  ca  thal
  15   5  ca  thal
  16   1  ca  thal
  16   2  ca  thal
  16   3  ca  thal
  16   4  ca  thal
  16   5  ca  thal
  17   1  ca  thal
  17   2  ca  thal
  17   3  ca  thal
  17   4  ca  thal
  17   5  ca  thal
  18   1  ca  thal
  18   2  ca  thal
  18   3  ca  thal
  18   4  ca  thal
  18   5  ca  thal
  19   1  ca  thal
  19   2  ca  thal
  19   3  ca  thal
  19   4  ca  thal
  19   5  ca  thal
  20   1  ca  thal
  20   2  ca  thal
  20   3  ca  thal
  20   4  ca  thal
  20   5  ca  thal
  21   1  ca  thal
  21   2  ca  thal
  21   3  ca  thal
  21   4  ca  thal
  21   5  ca  thal
  22   1  ca  thal
  22   2  ca  thal
  22   3  ca  thal
  22   4  ca  thal
  22   5  ca  thal
  23   1  ca  thal
  23   2  ca  thal
  23   3  ca  thal
  23   4  ca  thal
  23   5  ca  thal
  24   1  ca  thal
  24   2  ca  thal
  24   3  ca  thal
  24   4  ca  thal
  24   5  ca  thal
  25   1  ca  thal
  25   2  ca  thal
  25   3  ca  thal
  25   4  ca  thal
  25   5  ca  thal
  26   1  ca  thal
  26   2  ca  thal
  26   3  ca  thal
  26   4  ca  thal
  26   5  ca  thal
  27   1  ca  thal
  27   2  ca  thal
  27   3  ca  thal
  27   4  ca  thal
  27   5  ca  thal
  28   1  ca  thal
  28   2  ca  thal
  28   3  ca  thal
  28   4  ca  thal
  28   5  ca  thal
  29   1  ca  thal
  29   2  ca  thal
  29   3  ca  thal
  29   4  ca  thal
  29   5  ca  thal
  30   1  ca  thal
  30   2  ca  thal
  30   3  ca  thal
  30   4  ca  thal
  30   5  ca  thal
  31   1  ca  thal
  31   2  ca  thal
  31   3  ca  thal
  31   4  ca  thal
  31   5  ca  thal
  32   1  ca  thal
  32   2  ca  thal
  32   3  ca  thal
  32   4  ca  thal
  32   5  ca  thal
  33   1  ca  thal
  33   2  ca  thal
  33   3  ca  thal
  33   4  ca  thal
  33   5  ca  thal
  34   1  ca  thal
  34   2  ca  thal
  34   3  ca  thal
  34   4  ca  thal
  34   5  ca  thal
  35   1  ca  thal
  35   2  ca  thal
  35   3  ca  thal
  35   4  ca  thal
  35   5  ca  thal
  36   1  ca  thal
  36   2  ca  thal
  36   3  ca  thal
  36   4  ca  thal
  36   5  ca  thal
  37   1  ca  thal
  37   2  ca  thal
  37   3  ca  thal
  37   4  ca  thal
  37   5  ca  thal
  38   1  ca  thal
  38   2  ca  thal
  38   3  ca  thal
  38   4  ca  thal
  38   5  ca  thal
  39   1  ca  thal
  39   2  ca  thal
  39   3  ca  thal
  39   4  ca  thal
  39   5  ca  thal
  40   1  ca  thal
  40   2  ca  thal
  40   3  ca  thal
  40   4  ca  thal
  40   5  ca  thal
  41   1  ca  thal
  41   2  ca  thal
  41   3  ca  thal
  41   4  ca  thal
  41   5  ca  thal
  42   1  ca  thal
  42   2  ca  thal
  42   3  ca  thal
  42   4  ca  thal
  42   5  ca  thal
  43   1  ca  thal
  43   2  ca  thal
  43   3  ca  thal
  43   4  ca  thal
  43   5  ca  thal
  44   1  ca  thal
  44   2  ca  thal
  44   3  ca  thal
  44   4  ca  thal
  44   5  ca  thal
  45   1  ca  thal
  45   2  ca  thal
  45   3  ca  thal
  45   4  ca  thal
  45   5  ca  thal
  46   1  ca  thal
  46   2  ca  thal
  46   3  ca  thal
  46   4  ca  thal
  46   5  ca  thal
  47   1  ca  thal
  47   2  ca  thal
  47   3  ca  thal
  47   4  ca  thal
  47   5  ca  thal
  48   1  ca  thal
  48   2  ca  thal
  48   3  ca  thal
  48   4  ca  thal
  48   5  ca  thal
  49   1  ca  thal
  49   2  ca  thal
  49   3  ca  thal
  49   4  ca  thal
  49   5  ca  thal
  50   1  ca  thal
  50   2  ca  thal
  50   3  ca  thal
  50   4  ca  thal
  50   5  ca  thal
  51   1  ca  thal
  51   2  ca  thal
  51   3  ca  thal
  51   4  ca  thal
  51   5  ca  thal
  52   1  ca  thal
  52   2  ca  thal
  52   3  ca  thal
  52   4  ca  thal
  52   5  ca  thal
  53   1  ca  thal
  53   2  ca  thal
  53   3  ca  thal
  53   4  ca  thal
  53   5  ca  thal
  54   1  ca  thal
  54   2  ca  thal
  54   3  ca  thal
  54   4  ca  thal
  54   5  ca  thal
  55   1  ca  thal
  55   2  ca  thal
  55   3  ca  thal
  55   4  ca  thal
  55   5  ca  thal
  56   1  ca  thal
  56   2  ca  thal
  56   3  ca  thal
  56   4  ca  thal
  56   5  ca  thal
  57   1  ca  thal
  57   2  ca  thal
  57   3  ca  thal
  57   4  ca  thal
  57   5  ca  thal
  58   1  ca  thal
  58   2  ca  thal
  58   3  ca  thal
  58   4  ca  thal
  58   5  ca  thal
  59   1  ca  thal
  59   2  ca  thal
  59   3  ca  thal
  59   4  ca  thal
  59   5  ca  thal
  60   1  ca  thal
  60   2  ca  thal
  60   3  ca  thal
  60   4  ca  thal
  60   5  ca  thal
  61   1  ca  thal
  61   2  ca  thal
  61   3  ca  thal
  61   4  ca  thal
  61   5  ca  thal
  62   1  ca  thal
  62   2  ca  thal
  62   3  ca  thal
  62   4  ca  thal
  62   5  ca  thal
  63   1  ca  thal
  63   2  ca  thal
  63   3  ca  thal
  63   4  ca  thal
  63   5  ca  thal
  64   1  ca  thal
  64   2  ca  thal
  64   3  ca  thal
  64   4  ca  thal
  64   5  ca  thal
  65   1  ca  thal
  65   2  ca  thal
  65   3  ca  thal
  65   4  ca  thal
  65   5  ca  thal
  66   1  ca  thal
  66   2  ca  thal
  66   3  ca  thal
  66   4  ca  thal
  66   5  ca  thal
  67   1  ca  thal
  67   2  ca  thal
  67   3  ca  thal
  67   4  ca  thal
  67   5  ca  thal
  68   1  ca  thal
  68   2  ca  thal
  68   3  ca  thal
  68   4  ca  thal
  68   5  ca  thal
  69   1  ca  thal
  69   2  ca  thal
  69   3  ca  thal
  69   4  ca  thal
  69   5  ca  thal
  70   1  ca  thal
  70   2  ca  thal
  70   3  ca  thal
  70   4  ca  thal
  70   5  ca  thal
  71   1  ca  thal
  71   2  ca  thal
  71   3  ca  thal
  71   4  ca  thal
  71   5  ca  thal
  72   1  ca  thal
  72   2  ca  thal
  72   3  ca  thal
  72   4  ca  thal
  72   5  ca  thal
  73   1  ca  thal
  73   2  ca  thal
  73   3  ca  thal
  73   4  ca  thal
  73   5  ca  thal
  74   1  ca  thal
  74   2  ca  thal
  74   3  ca  thal
  74   4  ca  thal
  74   5  ca  thal
  75   1  ca  thal
  75   2  ca  thal
  75   3  ca  thal
  75   4  ca  thal
  75   5  ca  thal
  76   1  ca  thal
  76   2  ca  thal
  76   3  ca  thal
  76   4  ca  thal
  76   5  ca  thal
  77   1  ca  thal
  77   2  ca  thal
  77   3  ca  thal
  77   4  ca  thal
  77   5  ca  thal
  78   1  ca  thal
  78   2  ca  thal
  78   3  ca  thal
  78   4  ca  thal
  78   5  ca  thal
  79   1  ca  thal
  79   2  ca  thal
  79   3  ca  thal
  79   4  ca  thal
  79   5  ca  thal
  80   1  ca  thal
  80   2  ca  thal
  80   3  ca  thal
  80   4  ca  thal
  80   5  ca  thal
  81   1  ca  thal
  81   2  ca  thal
  81   3  ca  thal
  81   4  ca  thal
  81   5  ca  thal
  82   1  ca  thal
  82   2  ca  thal
  82   3  ca  thal
  82   4  ca  thal
  82   5  ca  thal
  83   1  ca  thal
  83   2  ca  thal
  83   3  ca  thal
  83   4  ca  thal
  83   5  ca  thal
  84   1  ca  thal
  84   2  ca  thal
  84   3  ca  thal
  84   4  ca  thal
  84   5  ca  thal
  85   1  ca  thal
  85   2  ca  thal
  85   3  ca  thal
  85   4  ca  thal
  85   5  ca  thal
  86   1  ca  thal
  86   2  ca  thal
  86   3  ca  thal
  86   4  ca  thal
  86   5  ca  thal
  87   1  ca  thal
  87   2  ca  thal
  87   3  ca  thal
  87   4  ca  thal
  87   5  ca  thal
  88   1  ca  thal
  88   2  ca  thal
  88   3  ca  thal
  88   4  ca  thal
  88   5  ca  thal
  89   1  ca  thal
  89   2  ca  thal
  89   3  ca  thal
  89   4  ca  thal
  89   5  ca  thal
  90   1  ca  thal
  90   2  ca  thal
  90   3  ca  thal
  90   4  ca  thal
  90   5  ca  thal
  91   1  ca  thal
  91   2  ca  thal
  91   3  ca  thal
  91   4  ca  thal
  91   5  ca  thal
  92   1  ca  thal
  92   2  ca  thal
  92   3  ca  thal
  92   4  ca  thal
  92   5  ca  thal
  93   1  ca  thal
  93   2  ca  thal
  93   3  ca  thal
  93   4  ca  thal
  93   5  ca  thal
  94   1  ca  thal
  94   2  ca  thal
  94   3  ca  thal
  94   4  ca  thal
  94   5  ca  thal
  95   1  ca  thal
  95   2  ca  thal
  95   3  ca  thal
  95   4  ca  thal
  95   5  ca  thal
  96   1  ca  thal
  96   2  ca  thal
  96   3  ca  thal
  96   4  ca  thal
  96   5  ca  thal
  97   1  ca  thal
  97   2  ca  thal
  97   3  ca  thal
  97   4  ca  thal
  97   5  ca  thal
  98   1  ca  thal
  98   2  ca  thal
  98   3  ca  thal
  98   4  ca  thal
  98   5  ca  thal
  99   1  ca  thal
  99   2  ca  thal
  99   3  ca  thal
  99   4  ca  thal
  99   5  ca  thal
  100   1  ca  thal
  100   2  ca  thal
  100   3  ca  thal
  100   4  ca  thal
  100   5  ca  thal</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>dataset <span class="ot">&lt;-</span> <span class="fu">complete</span>(imputed_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># Convert ca and thal to factors with predefined levels</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>dataset<span class="sc">$</span>ca <span class="ot">&lt;-</span> <span class="fu">factor</span>(dataset<span class="sc">$</span>ca, <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>))</span>
<span id="cb8-3"><a href="#cb8-3"></a>dataset<span class="sc">$</span>thal <span class="ot">&lt;-</span> <span class="fu">factor</span>(dataset<span class="sc">$</span>thal, <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">6</span>, <span class="dv">7</span>))</span>
<span id="cb8-4"><a href="#cb8-4"></a>methods <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">""</span>,<span class="st">"pmm"</span>,<span class="st">"polyreg"</span>,<span class="st">""</span>,<span class="st">""</span>)</span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="fu">names</span>(methods) <span class="ot">&lt;-</span> <span class="fu">names</span>(dataset)</span>
<span id="cb8-6"><a href="#cb8-6"></a>imputed_data <span class="ot">&lt;-</span> <span class="fu">mice</span>(dataset, <span class="at">method =</span> methods, <span class="at">m =</span> <span class="dv">5</span>, <span class="at">maxit =</span> <span class="dv">100</span>, <span class="at">seed =</span> <span class="dv">123</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 iter imp variable
  1   1
  1   2
  1   3
  1   4
  1   5
  2   1
  2   2
  2   3
  2   4
  2   5
  3   1
  3   2
  3   3
  3   4
  3   5
  4   1
  4   2
  4   3
  4   4
  4   5
  5   1
  5   2
  5   3
  5   4
  5   5
  6   1
  6   2
  6   3
  6   4
  6   5
  7   1
  7   2
  7   3
  7   4
  7   5
  8   1
  8   2
  8   3
  8   4
  8   5
  9   1
  9   2
  9   3
  9   4
  9   5
  10   1
  10   2
  10   3
  10   4
  10   5
  11   1
  11   2
  11   3
  11   4
  11   5
  12   1
  12   2
  12   3
  12   4
  12   5
  13   1
  13   2
  13   3
  13   4
  13   5
  14   1
  14   2
  14   3
  14   4
  14   5
  15   1
  15   2
  15   3
  15   4
  15   5
  16   1
  16   2
  16   3
  16   4
  16   5
  17   1
  17   2
  17   3
  17   4
  17   5
  18   1
  18   2
  18   3
  18   4
  18   5
  19   1
  19   2
  19   3
  19   4
  19   5
  20   1
  20   2
  20   3
  20   4
  20   5
  21   1
  21   2
  21   3
  21   4
  21   5
  22   1
  22   2
  22   3
  22   4
  22   5
  23   1
  23   2
  23   3
  23   4
  23   5
  24   1
  24   2
  24   3
  24   4
  24   5
  25   1
  25   2
  25   3
  25   4
  25   5
  26   1
  26   2
  26   3
  26   4
  26   5
  27   1
  27   2
  27   3
  27   4
  27   5
  28   1
  28   2
  28   3
  28   4
  28   5
  29   1
  29   2
  29   3
  29   4
  29   5
  30   1
  30   2
  30   3
  30   4
  30   5
  31   1
  31   2
  31   3
  31   4
  31   5
  32   1
  32   2
  32   3
  32   4
  32   5
  33   1
  33   2
  33   3
  33   4
  33   5
  34   1
  34   2
  34   3
  34   4
  34   5
  35   1
  35   2
  35   3
  35   4
  35   5
  36   1
  36   2
  36   3
  36   4
  36   5
  37   1
  37   2
  37   3
  37   4
  37   5
  38   1
  38   2
  38   3
  38   4
  38   5
  39   1
  39   2
  39   3
  39   4
  39   5
  40   1
  40   2
  40   3
  40   4
  40   5
  41   1
  41   2
  41   3
  41   4
  41   5
  42   1
  42   2
  42   3
  42   4
  42   5
  43   1
  43   2
  43   3
  43   4
  43   5
  44   1
  44   2
  44   3
  44   4
  44   5
  45   1
  45   2
  45   3
  45   4
  45   5
  46   1
  46   2
  46   3
  46   4
  46   5
  47   1
  47   2
  47   3
  47   4
  47   5
  48   1
  48   2
  48   3
  48   4
  48   5
  49   1
  49   2
  49   3
  49   4
  49   5
  50   1
  50   2
  50   3
  50   4
  50   5
  51   1
  51   2
  51   3
  51   4
  51   5
  52   1
  52   2
  52   3
  52   4
  52   5
  53   1
  53   2
  53   3
  53   4
  53   5
  54   1
  54   2
  54   3
  54   4
  54   5
  55   1
  55   2
  55   3
  55   4
  55   5
  56   1
  56   2
  56   3
  56   4
  56   5
  57   1
  57   2
  57   3
  57   4
  57   5
  58   1
  58   2
  58   3
  58   4
  58   5
  59   1
  59   2
  59   3
  59   4
  59   5
  60   1
  60   2
  60   3
  60   4
  60   5
  61   1
  61   2
  61   3
  61   4
  61   5
  62   1
  62   2
  62   3
  62   4
  62   5
  63   1
  63   2
  63   3
  63   4
  63   5
  64   1
  64   2
  64   3
  64   4
  64   5
  65   1
  65   2
  65   3
  65   4
  65   5
  66   1
  66   2
  66   3
  66   4
  66   5
  67   1
  67   2
  67   3
  67   4
  67   5
  68   1
  68   2
  68   3
  68   4
  68   5
  69   1
  69   2
  69   3
  69   4
  69   5
  70   1
  70   2
  70   3
  70   4
  70   5
  71   1
  71   2
  71   3
  71   4
  71   5
  72   1
  72   2
  72   3
  72   4
  72   5
  73   1
  73   2
  73   3
  73   4
  73   5
  74   1
  74   2
  74   3
  74   4
  74   5
  75   1
  75   2
  75   3
  75   4
  75   5
  76   1
  76   2
  76   3
  76   4
  76   5
  77   1
  77   2
  77   3
  77   4
  77   5
  78   1
  78   2
  78   3
  78   4
  78   5
  79   1
  79   2
  79   3
  79   4
  79   5
  80   1
  80   2
  80   3
  80   4
  80   5
  81   1
  81   2
  81   3
  81   4
  81   5
  82   1
  82   2
  82   3
  82   4
  82   5
  83   1
  83   2
  83   3
  83   4
  83   5
  84   1
  84   2
  84   3
  84   4
  84   5
  85   1
  85   2
  85   3
  85   4
  85   5
  86   1
  86   2
  86   3
  86   4
  86   5
  87   1
  87   2
  87   3
  87   4
  87   5
  88   1
  88   2
  88   3
  88   4
  88   5
  89   1
  89   2
  89   3
  89   4
  89   5
  90   1
  90   2
  90   3
  90   4
  90   5
  91   1
  91   2
  91   3
  91   4
  91   5
  92   1
  92   2
  92   3
  92   4
  92   5
  93   1
  93   2
  93   3
  93   4
  93   5
  94   1
  94   2
  94   3
  94   4
  94   5
  95   1
  95   2
  95   3
  95   4
  95   5
  96   1
  96   2
  96   3
  96   4
  96   5
  97   1
  97   2
  97   3
  97   4
  97   5
  98   1
  98   2
  98   3
  98   4
  98   5
  99   1
  99   2
  99   3
  99   4
  99   5
  100   1
  100   2
  100   3
  100   4
  100   5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>complete_data <span class="ot">&lt;-</span> <span class="fu">complete</span>(imputed_data)</span>
<span id="cb10-2"><a href="#cb10-2"></a></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co">#Check patient ID for duplicates</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="fu">any</span>(<span class="fu">duplicated</span>(complete_data<span class="sc">$</span>patient_id))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># Understand the structure of the data</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="fu">str</span>(complete_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   303 obs. of  15 variables:
 $ age       : int  63 67 67 37 41 56 62 57 63 53 ...
 $ sex       : int  1 1 1 1 0 1 0 0 1 1 ...
 $ cp        : int  1 4 4 3 2 2 4 4 4 4 ...
 $ trestbps  : int  145 160 120 130 130 120 140 120 130 140 ...
 $ chol      : int  233 286 229 250 204 236 268 354 254 203 ...
 $ fbs       : int  1 0 0 0 0 0 0 0 0 1 ...
 $ restecg   : int  2 2 2 0 2 0 2 0 2 2 ...
 $ thalach   : int  150 108 129 187 172 178 160 163 147 155 ...
 $ exang     : int  0 1 1 0 0 0 0 1 0 1 ...
 $ oldpeak   : num  2.3 1.5 2.6 3.5 1.4 0.8 3.6 0.6 1.4 3.1 ...
 $ slope     : int  3 2 2 3 1 1 3 1 2 3 ...
 $ ca        : Factor w/ 4 levels "0","1","2","3": 1 4 3 1 1 1 3 1 2 1 ...
 $ thal      : Factor w/ 3 levels "3","6","7": 2 1 3 1 1 1 1 1 3 3 ...
 $ num       : int  0 2 1 0 0 0 3 0 2 1 ...
 $ patient_id: int  1 2 3 4 5 6 7 8 9 10 ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="fu">summary</span>(complete_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      age             sex               cp           trestbps    
 Min.   :29.00   Min.   :0.0000   Min.   :1.000   Min.   : 94.0  
 1st Qu.:48.00   1st Qu.:0.0000   1st Qu.:3.000   1st Qu.:120.0  
 Median :56.00   Median :1.0000   Median :3.000   Median :130.0  
 Mean   :54.44   Mean   :0.6799   Mean   :3.158   Mean   :131.7  
 3rd Qu.:61.00   3rd Qu.:1.0000   3rd Qu.:4.000   3rd Qu.:140.0  
 Max.   :77.00   Max.   :1.0000   Max.   :4.000   Max.   :200.0  
      chol            fbs            restecg          thalach     
 Min.   :126.0   Min.   :0.0000   Min.   :0.0000   Min.   : 71.0  
 1st Qu.:211.0   1st Qu.:0.0000   1st Qu.:0.0000   1st Qu.:133.5  
 Median :241.0   Median :0.0000   Median :1.0000   Median :153.0  
 Mean   :246.7   Mean   :0.1485   Mean   :0.9901   Mean   :149.6  
 3rd Qu.:275.0   3rd Qu.:0.0000   3rd Qu.:2.0000   3rd Qu.:166.0  
 Max.   :564.0   Max.   :1.0000   Max.   :2.0000   Max.   :202.0  
     exang           oldpeak         slope       ca      thal   
 Min.   :0.0000   Min.   :0.00   Min.   :1.000   0:179   3:167  
 1st Qu.:0.0000   1st Qu.:0.00   1st Qu.:1.000   1: 66   6: 18  
 Median :0.0000   Median :0.80   Median :2.000   2: 38   7:118  
 Mean   :0.3267   Mean   :1.04   Mean   :1.601   3: 20          
 3rd Qu.:1.0000   3rd Qu.:1.60   3rd Qu.:2.000                  
 Max.   :1.0000   Max.   :6.20   Max.   :3.000                  
      num           patient_id   
 Min.   :0.0000   Min.   :  1.0  
 1st Qu.:0.0000   1st Qu.: 76.5  
 Median :0.0000   Median :152.0  
 Mean   :0.9373   Mean   :152.0  
 3rd Qu.:2.0000   3rd Qu.:227.5  
 Max.   :4.0000   Max.   :303.0  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co">#Histograms to investigate distribution of numeric variables</span></span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="fu">hist</span>(complete_data<span class="sc">$</span>age, </span>
<span id="cb16-3"><a href="#cb16-3"></a>     <span class="at">main =</span> <span class="st">"Age Distribution"</span>, </span>
<span id="cb16-4"><a href="#cb16-4"></a>     <span class="at">xlab =</span> <span class="st">"Age"</span>, </span>
<span id="cb16-5"><a href="#cb16-5"></a>     <span class="at">col =</span> <span class="st">"lightblue"</span>, </span>
<span id="cb16-6"><a href="#cb16-6"></a>     <span class="at">border =</span> <span class="st">"black"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="fu">hist</span>(complete_data<span class="sc">$</span>trestbps, </span>
<span id="cb17-2"><a href="#cb17-2"></a>     <span class="at">main =</span> <span class="st">"Resting Blood Pressure Distribution"</span>, </span>
<span id="cb17-3"><a href="#cb17-3"></a>     <span class="at">xlab =</span> <span class="st">"Blood Pressure (mm Hg)"</span>, </span>
<span id="cb17-4"><a href="#cb17-4"></a>     <span class="at">col =</span> <span class="st">"lightgreen"</span>, </span>
<span id="cb17-5"><a href="#cb17-5"></a>     <span class="at">border =</span> <span class="st">"black"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="fu">hist</span>(complete_data<span class="sc">$</span>chol, </span>
<span id="cb18-2"><a href="#cb18-2"></a>     <span class="at">main =</span> <span class="st">"Cholesterol Distribution"</span>, </span>
<span id="cb18-3"><a href="#cb18-3"></a>     <span class="at">xlab =</span> <span class="st">"Cholesterol (mg/dl)"</span>, </span>
<span id="cb18-4"><a href="#cb18-4"></a>     <span class="at">col =</span> <span class="st">"orange"</span>, </span>
<span id="cb18-5"><a href="#cb18-5"></a>     <span class="at">border =</span> <span class="st">"black"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="fu">hist</span>(complete_data<span class="sc">$</span>thalach, </span>
<span id="cb19-2"><a href="#cb19-2"></a>     <span class="at">main =</span> <span class="st">"Max Heart Rate Distribution"</span>, </span>
<span id="cb19-3"><a href="#cb19-3"></a>     <span class="at">xlab =</span> <span class="st">"Heart Rate (bpm)"</span>, </span>
<span id="cb19-4"><a href="#cb19-4"></a>     <span class="at">col =</span> <span class="st">"purple"</span>, </span>
<span id="cb19-5"><a href="#cb19-5"></a>     <span class="at">border =</span> <span class="st">"black"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(complete_data<span class="sc">$</span>oldpeak, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), </span>
<span id="cb20-2"><a href="#cb20-2"></a>     <span class="at">main =</span> <span class="st">"Density of ST Depression (Oldpeak)"</span>, </span>
<span id="cb20-3"><a href="#cb20-3"></a>     <span class="at">xlab =</span> <span class="st">"Oldpeak"</span>, </span>
<span id="cb20-4"><a href="#cb20-4"></a>     <span class="at">col =</span> <span class="st">"red"</span>, </span>
<span id="cb20-5"><a href="#cb20-5"></a>     <span class="at">lwd =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co">#Barplots for categorical variables that show the frequency of each category</span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="fu">barplot</span>(<span class="fu">table</span>(complete_data<span class="sc">$</span>sex), </span>
<span id="cb21-3"><a href="#cb21-3"></a>        <span class="at">main =</span> <span class="st">"Sex Distribution"</span>, </span>
<span id="cb21-4"><a href="#cb21-4"></a>        <span class="at">xlab =</span> <span class="st">"Sex (1 = Male, 0 = Female)"</span>, </span>
<span id="cb21-5"><a href="#cb21-5"></a>        <span class="at">ylab =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb21-6"><a href="#cb21-6"></a>        <span class="at">col =</span> <span class="st">"lightblue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-6.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(complete_data<span class="sc">$</span>cp), </span>
<span id="cb22-2"><a href="#cb22-2"></a>        <span class="at">main =</span> <span class="st">"Chest Pain Type Distribution"</span>, </span>
<span id="cb22-3"><a href="#cb22-3"></a>        <span class="at">xlab =</span> <span class="st">"Chest Pain Type"</span>, </span>
<span id="cb22-4"><a href="#cb22-4"></a>        <span class="at">ylab =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb22-5"><a href="#cb22-5"></a>        <span class="at">col =</span> <span class="st">"lightgreen"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-7.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(complete_data<span class="sc">$</span>fbs), </span>
<span id="cb23-2"><a href="#cb23-2"></a>        <span class="at">main =</span> <span class="st">"Fasting Blood Sugar Distribution"</span>, </span>
<span id="cb23-3"><a href="#cb23-3"></a>        <span class="at">xlab =</span> <span class="st">"Fasting Blood Sugar (&gt;120 mg/dl)"</span>, </span>
<span id="cb23-4"><a href="#cb23-4"></a>        <span class="at">ylab =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb23-5"><a href="#cb23-5"></a>        <span class="at">col =</span> <span class="st">"orange"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-8.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(complete_data<span class="sc">$</span>restecg), </span>
<span id="cb24-2"><a href="#cb24-2"></a>        <span class="at">main =</span> <span class="st">"Resting ECG Distribution"</span>, </span>
<span id="cb24-3"><a href="#cb24-3"></a>        <span class="at">xlab =</span> <span class="st">"Resting ECG"</span>, </span>
<span id="cb24-4"><a href="#cb24-4"></a>        <span class="at">ylab =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb24-5"><a href="#cb24-5"></a>        <span class="at">col =</span> <span class="st">"purple"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-9.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(complete_data<span class="sc">$</span>exang), </span>
<span id="cb25-2"><a href="#cb25-2"></a>        <span class="at">main =</span> <span class="st">"Exercise-Induced Angina Distribution"</span>, </span>
<span id="cb25-3"><a href="#cb25-3"></a>        <span class="at">xlab =</span> <span class="st">"Exercise-Induced Angina (1 = Yes, 0 = No)"</span>, </span>
<span id="cb25-4"><a href="#cb25-4"></a>        <span class="at">ylab =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb25-5"><a href="#cb25-5"></a>        <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-10.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(complete_data<span class="sc">$</span>slope), </span>
<span id="cb26-2"><a href="#cb26-2"></a>        <span class="at">main =</span> <span class="st">"Slope of Peak Exercise Distribution"</span>, </span>
<span id="cb26-3"><a href="#cb26-3"></a>        <span class="at">xlab =</span> <span class="st">"Slope"</span>, </span>
<span id="cb26-4"><a href="#cb26-4"></a>        <span class="at">ylab =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb26-5"><a href="#cb26-5"></a>        <span class="at">col =</span> <span class="st">"cyan"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-11.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(complete_data<span class="sc">$</span>ca), </span>
<span id="cb27-2"><a href="#cb27-2"></a>        <span class="at">main =</span> <span class="st">"Number of Major Vessels Distribution"</span>, </span>
<span id="cb27-3"><a href="#cb27-3"></a>        <span class="at">xlab =</span> <span class="st">"Number of Vessels"</span>, </span>
<span id="cb27-4"><a href="#cb27-4"></a>        <span class="at">ylab =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb27-5"><a href="#cb27-5"></a>        <span class="at">col =</span> <span class="st">"pink"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-12.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(complete_data<span class="sc">$</span>thal), </span>
<span id="cb28-2"><a href="#cb28-2"></a>        <span class="at">main =</span> <span class="st">"Thalassemia Distribution"</span>, </span>
<span id="cb28-3"><a href="#cb28-3"></a>        <span class="at">xlab =</span> <span class="st">"Thalassemia"</span>, </span>
<span id="cb28-4"><a href="#cb28-4"></a>        <span class="at">ylab =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb28-5"><a href="#cb28-5"></a>        <span class="at">col =</span> <span class="st">"purple"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-13.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(complete_data<span class="sc">$</span>num), </span>
<span id="cb29-2"><a href="#cb29-2"></a>        <span class="at">main =</span> <span class="st">"Disease Outcome Distribution"</span>, </span>
<span id="cb29-3"><a href="#cb29-3"></a>        <span class="at">xlab =</span> <span class="st">"Disease Outcome"</span>, </span>
<span id="cb29-4"><a href="#cb29-4"></a>        <span class="at">ylab =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb29-5"><a href="#cb29-5"></a>        <span class="at">col =</span> <span class="st">"orange"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-3-14.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Insights from the Exploratory Data Analysis (EDA)</p>
<p>The EDA on the heart disease dataset involved preprocessing, handling missing values, and visualizing key attributes to understand their distributions. The dataset was cleaned by converting categorical variables into factors, imputing missing values using the mice package, and checking for duplicates.</p>
<p>Key Insights: Age Distribution: The dataset primarily consists of middle-aged individuals, with most patients between 50 and 60 years old. Blood Pressure &amp; Cholesterol: Resting blood pressure is mostly between 100 and 140 mm Hg, while cholesterol levels show a right-skewed distribution, with many values between 200-300 mg/dL and some extreme cases. Max Heart Rate: The majority of patients have a max heart rate between 120 and 170 bpm, which follows a near-normal distribution. Chest Pain Types: Most patients experience asymptomatic chest pain (type 4), indicating that many do not have typical angina symptoms. Fasting Blood Sugar &amp; ECG: Most patients have normal fasting blood sugar, and ECG results show that normal readings (category 0) are the most common. Number of Major Vessels &amp; Disease Outcome: The majority of patients have no major vessels affected (ca = 0), and most do not have severe heart disease (num = 0), making it an imbalanced dataset.</p>
</section>
</section>
<section id="create-visualizations-in-order-to-show-which-variables-seem-to-be-more-associated-with-heart-disease" class="level2">
<h2 class="anchored" data-anchor-id="create-visualizations-in-order-to-show-which-variables-seem-to-be-more-associated-with-heart-disease">1.2. Create visualizations in order to show which variables seem to be more associated with heart disease</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="co">#num variable represents heart disease outcomes and thus we can analyze how each variable relates to it</span></span>
<span id="cb30-2"><a href="#cb30-2"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb30-3"><a href="#cb30-3"></a></span>
<span id="cb30-4"><a href="#cb30-4"></a><span class="co">#Sex vs. Heart Disease</span></span>
<span id="cb30-5"><a href="#cb30-5"></a><span class="fu">ggplot</span>(complete_data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(sex), <span class="at">fill =</span> <span class="fu">factor</span>(num))) <span class="sc">+</span></span>
<span id="cb30-6"><a href="#cb30-6"></a>  <span class="fu">geom_bar</span>(<span class="at">position =</span> <span class="st">"fill"</span>) <span class="sc">+</span></span>
<span id="cb30-7"><a href="#cb30-7"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Proportion of Heart Disease by Sex"</span>, </span>
<span id="cb30-8"><a href="#cb30-8"></a>       <span class="at">x =</span> <span class="st">"Sex (1 = Male, 0 = Female)"</span>, </span>
<span id="cb30-9"><a href="#cb30-9"></a>       <span class="at">y =</span> <span class="st">"Proportion"</span>, </span>
<span id="cb30-10"><a href="#cb30-10"></a>       <span class="at">fill =</span> <span class="st">"Heart Disease"</span>) <span class="sc">+</span></span>
<span id="cb30-11"><a href="#cb30-11"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="co">#Chest pain type vs. Heart Disease</span></span>
<span id="cb31-2"><a href="#cb31-2"></a><span class="fu">ggplot</span>(complete_data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(cp), <span class="at">fill =</span> <span class="fu">factor</span>(num))) <span class="sc">+</span></span>
<span id="cb31-3"><a href="#cb31-3"></a>  <span class="fu">geom_bar</span>(<span class="at">position =</span> <span class="st">"fill"</span>) <span class="sc">+</span></span>
<span id="cb31-4"><a href="#cb31-4"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Proportion of Heart Disease by Chest Pain Type"</span>, </span>
<span id="cb31-5"><a href="#cb31-5"></a>       <span class="at">x =</span> <span class="st">"Chest Pain Type"</span>, </span>
<span id="cb31-6"><a href="#cb31-6"></a>       <span class="at">y =</span> <span class="st">"Proportion"</span>, </span>
<span id="cb31-7"><a href="#cb31-7"></a>       <span class="at">fill =</span> <span class="st">"Heart Disease"</span>) <span class="sc">+</span></span>
<span id="cb31-8"><a href="#cb31-8"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="co">#Numeric variables</span></span>
<span id="cb32-2"><a href="#cb32-2"></a><span class="co">#Age vs. Heart Disease</span></span>
<span id="cb32-3"><a href="#cb32-3"></a><span class="fu">ggplot</span>(complete_data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(num), <span class="at">y =</span> age, <span class="at">fill =</span> <span class="fu">factor</span>(num))) <span class="sc">+</span></span>
<span id="cb32-4"><a href="#cb32-4"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb32-5"><a href="#cb32-5"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Age Distribution by Heart Disease Status"</span>, </span>
<span id="cb32-6"><a href="#cb32-6"></a>       <span class="at">x =</span> <span class="st">"Heart Disease Status"</span>, </span>
<span id="cb32-7"><a href="#cb32-7"></a>       <span class="at">y =</span> <span class="st">"Age"</span>, </span>
<span id="cb32-8"><a href="#cb32-8"></a>       <span class="at">fill =</span> <span class="st">"Heart Disease"</span>) <span class="sc">+</span></span>
<span id="cb32-9"><a href="#cb32-9"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="co">#Cholesterol vs. Heart Disease</span></span>
<span id="cb33-2"><a href="#cb33-2"></a><span class="fu">ggplot</span>(complete_data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(num), <span class="at">y =</span> chol, <span class="at">fill =</span> <span class="fu">factor</span>(num))) <span class="sc">+</span></span>
<span id="cb33-3"><a href="#cb33-3"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb33-4"><a href="#cb33-4"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Cholesterol Levels by Heart Disease Status"</span>, </span>
<span id="cb33-5"><a href="#cb33-5"></a>       <span class="at">x =</span> <span class="st">"Heart Disease Status"</span>, </span>
<span id="cb33-6"><a href="#cb33-6"></a>       <span class="at">y =</span> <span class="st">"Cholesterol"</span>, </span>
<span id="cb33-7"><a href="#cb33-7"></a>       <span class="at">fill =</span> <span class="st">"Heart Disease"</span>) <span class="sc">+</span></span>
<span id="cb33-8"><a href="#cb33-8"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="co">#Maximum Heart Rate vs. Heart Disease</span></span>
<span id="cb34-2"><a href="#cb34-2"></a><span class="fu">ggplot</span>(complete_data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(num), <span class="at">y =</span> thalach, <span class="at">fill =</span> <span class="fu">factor</span>(num))) <span class="sc">+</span></span>
<span id="cb34-3"><a href="#cb34-3"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb34-4"><a href="#cb34-4"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Max Heart Rate by Heart Disease Status"</span>, </span>
<span id="cb34-5"><a href="#cb34-5"></a>       <span class="at">x =</span> <span class="st">"Heart Disease Status"</span>, </span>
<span id="cb34-6"><a href="#cb34-6"></a>       <span class="at">y =</span> <span class="st">"Max Heart Rate"</span>, </span>
<span id="cb34-7"><a href="#cb34-7"></a>       <span class="at">fill =</span> <span class="st">"Heart Disease"</span>) <span class="sc">+</span></span>
<span id="cb34-8"><a href="#cb34-8"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="co">#Correlation Analysis</span></span>
<span id="cb35-2"><a href="#cb35-2"></a>cor_matrix <span class="ot">&lt;-</span> <span class="fu">cor</span>(complete_data[, <span class="fu">sapply</span>(complete_data, is.numeric)])</span>
<span id="cb35-3"><a href="#cb35-3"></a>cor_with_disease <span class="ot">&lt;-</span> cor_matrix[<span class="st">"num"</span>, ]</span>
<span id="cb35-4"><a href="#cb35-4"></a><span class="fu">print</span>(cor_with_disease)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        age         sex          cp    trestbps        chol         fbs 
 0.22285342  0.22446851  0.40707542  0.15775408  0.07090876  0.05918609 
    restecg     thalach       exang     oldpeak       slope         num 
 0.18369635 -0.41504046  0.39705730  0.50409197  0.37795660  1.00000000 
 patient_id 
 0.02925623 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="fu">barplot</span>(<span class="fu">abs</span>(cor_with_disease[<span class="sc">-</span><span class="fu">which</span>(<span class="fu">names</span>(cor_with_disease) <span class="sc">==</span> <span class="st">"num"</span>)]), </span>
<span id="cb37-2"><a href="#cb37-2"></a>        <span class="at">main =</span> <span class="st">"Correlation with Heart Disease"</span>, </span>
<span id="cb37-3"><a href="#cb37-3"></a>        <span class="at">xlab =</span> <span class="st">"Variables"</span>, </span>
<span id="cb37-4"><a href="#cb37-4"></a>        <span class="at">ylab =</span> <span class="st">"Correlation Coefficient"</span>, </span>
<span id="cb37-5"><a href="#cb37-5"></a>        <span class="at">col =</span> <span class="st">"blue"</span>, </span>
<span id="cb37-6"><a href="#cb37-6"></a>        <span class="at">las =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-6.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="co">#The variables with the highest correlation are cp, oldpeak, exang, thalach, slope, ca, age in this order.</span></span>
<span id="cb38-2"><a href="#cb38-2"></a><span class="co">#They must be investigated more</span></span>
<span id="cb38-3"><a href="#cb38-3"></a></span>
<span id="cb38-4"><a href="#cb38-4"></a><span class="co">#Violin plot for Max Heart Rate achieved vs. Heart Disease</span></span>
<span id="cb38-5"><a href="#cb38-5"></a><span class="fu">ggplot</span>(complete_data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(num), <span class="at">y =</span> thalach, <span class="at">fill =</span> <span class="fu">factor</span>(num))) <span class="sc">+</span></span>
<span id="cb38-6"><a href="#cb38-6"></a>  <span class="fu">geom_violin</span>(<span class="at">trim =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb38-7"><a href="#cb38-7"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.1</span>, <span class="at">outlier.color =</span> <span class="st">"red"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb38-8"><a href="#cb38-8"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Distribution of Max Heart Rate by Heart Disease Status"</span>, </span>
<span id="cb38-9"><a href="#cb38-9"></a>       <span class="at">x =</span> <span class="st">"Heart Disease Status"</span>, </span>
<span id="cb38-10"><a href="#cb38-10"></a>       <span class="at">y =</span> <span class="st">"Max Heart Rate"</span>) <span class="sc">+</span></span>
<span id="cb38-11"><a href="#cb38-11"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-7.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="fu">library</span>(ggridges)</span>
<span id="cb39-2"><a href="#cb39-2"></a><span class="fu">ggplot</span>(complete_data, <span class="fu">aes</span>(<span class="at">x =</span> oldpeak, <span class="at">y =</span> <span class="fu">factor</span>(num), <span class="at">fill =</span> <span class="fu">factor</span>(num))) <span class="sc">+</span></span>
<span id="cb39-3"><a href="#cb39-3"></a>  <span class="fu">geom_density_ridges</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb39-4"><a href="#cb39-4"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Distribution of ST Depression by Heart Disease Status"</span>, </span>
<span id="cb39-5"><a href="#cb39-5"></a>       <span class="at">x =</span> <span class="st">"ST Depression (Oldpeak)"</span>, </span>
<span id="cb39-6"><a href="#cb39-6"></a>       <span class="at">y =</span> <span class="st">"Heart Disease Status"</span>, </span>
<span id="cb39-7"><a href="#cb39-7"></a>       <span class="at">fill =</span> <span class="st">"Heart Disease"</span>) <span class="sc">+</span></span>
<span id="cb39-8"><a href="#cb39-8"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.504</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-8.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="fu">ggplot</span>(complete_data, <span class="fu">aes</span>(<span class="at">x =</span> oldpeak, <span class="at">y =</span> thalach, <span class="at">color =</span> <span class="fu">factor</span>(num))) <span class="sc">+</span></span>
<span id="cb41-2"><a href="#cb41-2"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb41-3"><a href="#cb41-3"></a>  <span class="fu">geom_density_2d</span>() <span class="sc">+</span></span>
<span id="cb41-4"><a href="#cb41-4"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Max Heart Rate vs. ST Depression by Heart Disease Status"</span>, </span>
<span id="cb41-5"><a href="#cb41-5"></a>       <span class="at">x =</span> <span class="st">"ST Depression (Oldpeak)"</span>, </span>
<span id="cb41-6"><a href="#cb41-6"></a>       <span class="at">y =</span> <span class="st">"Max Heart Rate"</span>, </span>
<span id="cb41-7"><a href="#cb41-7"></a>       <span class="at">color =</span> <span class="st">"Heart Disease Status"</span>) <span class="sc">+</span></span>
<span id="cb41-8"><a href="#cb41-8"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-9.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a><span class="co">#Pairwise scatterplots with num grouping</span></span>
<span id="cb42-2"><a href="#cb42-2"></a><span class="fu">library</span>(GGally)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 method overwritten by 'GGally':
  method from   
  +.gg   ggplot2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a><span class="fu">ggpairs</span>(complete_data, </span>
<span id="cb44-2"><a href="#cb44-2"></a>        <span class="at">columns =</span> <span class="fu">c</span>(<span class="st">"age"</span>, <span class="st">"trestbps"</span>, <span class="st">"chol"</span>, <span class="st">"thalach"</span>, <span class="st">"oldpeak"</span>), </span>
<span id="cb44-3"><a href="#cb44-3"></a>        <span class="fu">aes</span>(<span class="at">color =</span> <span class="fu">factor</span>(num), <span class="at">alpha =</span> <span class="fl">0.7</span>)) <span class="sc">+</span></span>
<span id="cb44-4"><a href="#cb44-4"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb44-5"><a href="#cb44-5"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Pairwise Relationships by Heart Disease Status"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-10.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a><span class="co">#Conditional Density for Age</span></span>
<span id="cb45-2"><a href="#cb45-2"></a><span class="fu">ggplot</span>(complete_data, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">fill =</span> <span class="fu">factor</span>(num))) <span class="sc">+</span></span>
<span id="cb45-3"><a href="#cb45-3"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb45-4"><a href="#cb45-4"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Age Distribution by Heart Disease Status"</span>, </span>
<span id="cb45-5"><a href="#cb45-5"></a>       <span class="at">x =</span> <span class="st">"Age"</span>, </span>
<span id="cb45-6"><a href="#cb45-6"></a>       <span class="at">y =</span> <span class="st">"Density"</span>, </span>
<span id="cb45-7"><a href="#cb45-7"></a>       <span class="at">fill =</span> <span class="st">"Heart Disease Status"</span>) <span class="sc">+</span></span>
<span id="cb45-8"><a href="#cb45-8"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-11.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a><span class="co">#Heatmap</span></span>
<span id="cb46-2"><a href="#cb46-2"></a></span>
<span id="cb46-3"><a href="#cb46-3"></a><span class="fu">library</span>(corrplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>corrplot 0.95 loaded</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1"></a>cor_matrix <span class="ot">&lt;-</span> <span class="fu">cor</span>(complete_data[, <span class="fu">c</span>(<span class="st">"age"</span>, <span class="st">"trestbps"</span>, <span class="st">"chol"</span>, <span class="st">"thalach"</span>, <span class="st">"oldpeak"</span>)], </span>
<span id="cb48-2"><a href="#cb48-2"></a>                  <span class="at">use =</span> <span class="st">"complete.obs"</span>)</span>
<span id="cb48-3"><a href="#cb48-3"></a><span class="fu">corrplot</span>(cor_matrix, <span class="at">method =</span> <span class="st">"color"</span>, <span class="at">addCoef.col =</span> <span class="st">"black"</span>, <span class="at">tl.col =</span> <span class="st">"black"</span>, </span>
<span id="cb48-4"><a href="#cb48-4"></a>         <span class="at">title =</span> <span class="st">"Correlation Between Variables"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-12.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="co">#Interaction Plot for oldpeak and thalach</span></span>
<span id="cb49-2"><a href="#cb49-2"></a><span class="fu">library</span>(interactions)</span>
<span id="cb49-3"><a href="#cb49-3"></a>interaction_model <span class="ot">&lt;-</span> <span class="fu">lm</span>(num <span class="sc">~</span> oldpeak <span class="sc">*</span> thalach, <span class="at">data =</span> complete_data)</span>
<span id="cb49-4"><a href="#cb49-4"></a><span class="fu">interact_plot</span>(interaction_model, <span class="at">pred =</span> oldpeak, <span class="at">modx =</span> thalach, </span>
<span id="cb49-5"><a href="#cb49-5"></a>              <span class="at">modx.values =</span> <span class="st">"terciles"</span>, <span class="at">plot.points =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb49-6"><a href="#cb49-6"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Interaction Between ST Depression and Max Heart Rate"</span>, </span>
<span id="cb49-7"><a href="#cb49-7"></a>       <span class="at">x =</span> <span class="st">"ST Depression (Oldpeak)"</span>, </span>
<span id="cb49-8"><a href="#cb49-8"></a>       <span class="at">y =</span> <span class="st">"Heart Disease Probability"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Medians of each tercile of thalach are 126, 154, 172</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-13.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="co">#PCA</span></span>
<span id="cb51-2"><a href="#cb51-2"></a><span class="fu">library</span>(ggfortify)</span>
<span id="cb51-3"><a href="#cb51-3"></a>pca_model <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(complete_data[, <span class="fu">sapply</span>(complete_data, is.numeric)], <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb51-4"><a href="#cb51-4"></a><span class="fu">autoplot</span>(pca_model, <span class="at">data =</span> complete_data, <span class="at">colour =</span> <span class="st">"num"</span>) <span class="sc">+</span></span>
<span id="cb51-5"><a href="#cb51-5"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"PCA of Heart Disease Dataset"</span>, </span>
<span id="cb51-6"><a href="#cb51-6"></a>       <span class="at">x =</span> <span class="st">"Principal Component 1"</span>, </span>
<span id="cb51-7"><a href="#cb51-7"></a>       <span class="at">y =</span> <span class="st">"Principal Component 2"</span>, </span>
<span id="cb51-8"><a href="#cb51-8"></a>       <span class="at">color =</span> <span class="st">"Heart Disease Status"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-4-14.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We created a big number of different plots in order to try and understand this dataset (some of them not so clear and matching with this data but we tried a lot just to understand).</p>
<p>Key Trends and Correlations with Heart Disease From all the visualizations and analyses, several strong trends emerge regarding factors most correlated with heart disease:</p>
<p>ST Depression (Oldpeak) is the strongest predictor</p>
<p>Higher oldpeak values (ST depression induced by exercise) are clearly associated with more severe heart disease. This makes sense because ST depression reflects ischemia, meaning that the heart is not getting enough oxygen during exertion. The interaction plots and scatterplots confirm that as oldpeak increases, the probability of heart disease rises, particularly when max heart rate is low.</p>
<p>Max Heart Rate (Thalach) is inversely correlated with heart disease</p>
<p>Patients with heart disease tend to have a lower max heart rate, while those without it reach higher heart rate levels. This makes sense physiologically, as a lower max heart rate indicates reduced cardiovascular efficiency and possible arterial blockages.</p>
<p>Chest Pain Type (CP) is highly associated with heart disease</p>
<p>Patients with asymptomatic chest pain (type 4) have the highest proportion of heart disease cases. Meanwhile, those with typical angina (type 1) have the lowest risk.</p>
<p>Exercise-Induced Angina (Exang) is another key predictor</p>
<p>Patients who experience exercise-induced angina are far more likely to have heart disease.</p>
<p>The Slope of the ST Segment (Slope) provides additional predictive power</p>
<p>Patients with a flat or downsloping ST segment (slope = 2 or 3) tend to have more severe heart disease. A downsloping ST segment is often a marker of ischemia and poor prognosis, making it an important feature in diagnostics.</p>
<p>Age is moderately correlated with heart disease severity</p>
<p>Older individuals tend to have higher disease severity, but some younger patients also show heart disease presence. This suggests that while age is a risk factor, it is not a definitive indicator—lifestyle, genetics, and other health conditions also play roles.</p>
<p>Cholesterol and Resting Blood Pressure (Trestbps) show weak correlation</p>
<p>Despite being common risk factors for cardiovascular disease, cholesterol and blood pressure do not strongly correlate with heart disease presence in this dataset. This could be due to the fact that cholesterol levels fluctuate based on diet and medication use, and resting blood pressure alone does not capture dynamic heart function under stress.</p>
</section>
</section>
<section id="difference-in-mortality-rates-in-hospitalized-covid-19-patients" class="level1">
<h1>2. Difference in mortality rates in hospitalized COVID-19 patients</h1>
<p>Using the supplementary material from the <a href="https://www.frontiersin.org/articles/10.3389/fmed.2022.987182/full">Difference in mortality rates in hospitalized COVID-19 patients identified by cytokine profile clustering using a machine learning approach: An outcome prediction alternative</a>, perform the following tasks</p>
<section id="reproduce-figure-1-from-the-publication" class="level2">
<h2 class="anchored" data-anchor-id="reproduce-figure-1-from-the-publication">2.1. Reproduce Figure 1 from the publication</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb52-2"><a href="#cb52-2"></a><span class="fu">library</span>(pheatmap) </span>
<span id="cb52-3"><a href="#cb52-3"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb52-4"><a href="#cb52-4"></a><span class="fu">library</span>(readxl)     </span>
<span id="cb52-5"><a href="#cb52-5"></a><span class="fu">library</span>(ComplexHeatmap)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: grid</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>========================================
ComplexHeatmap version 2.22.0
Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/
Github page: https://github.com/jokergoo/ComplexHeatmap
Documentation: http://jokergoo.github.io/ComplexHeatmap-reference

If you use it in published research, please cite either one:
- Gu, Z. Complex Heatmap Visualization. iMeta 2022.
- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional 
    genomic data. Bioinformatics 2016.


The new InteractiveComplexHeatmap package can directly export static 
complex heatmaps into an interactive Shiny app with zero effort. Have a try!

This message can be suppressed by:
  suppressPackageStartupMessages(library(ComplexHeatmap))
========================================
! pheatmap() has been masked by ComplexHeatmap::pheatmap(). Most of the arguments
   in the original pheatmap() are identically supported in the new function. You 
   can still use the original function by explicitly calling pheatmap::pheatmap().</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ComplexHeatmap'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:pheatmap':

    pheatmap</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a><span class="fu">library</span>(gridExtra)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'gridExtra'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    combine</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1"></a><span class="fu">library</span>(kableExtra)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'kableExtra'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    group_rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb63-2"><a href="#cb63-2"></a></span>
<span id="cb63-3"><a href="#cb63-3"></a><span class="co"># library(knitr)</span></span>
<span id="cb63-4"><a href="#cb63-4"></a></span>
<span id="cb63-5"><a href="#cb63-5"></a><span class="co"># Load clinical data from supplementary tables</span></span>
<span id="cb63-6"><a href="#cb63-6"></a>clinical_data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"data_article/Table 1.XLSX"</span>, <span class="at">skip=</span><span class="dv">1</span>)  </span>
<span id="cb63-7"><a href="#cb63-7"></a></span>
<span id="cb63-8"><a href="#cb63-8"></a><span class="co"># Check the structure of the data</span></span>
<span id="cb63-9"><a href="#cb63-9"></a><span class="fu">str</span>(clinical_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [214 × 18] (S3: tbl_df/tbl/data.frame)
 $ ID                     : chr [1:214] "COAG-HSJD-001" "COAG-HSJD-002" "COAG-HSJD-003" "COAG-HSJD-004" ...
 $ Age                    : num [1:214] 63 52 74 78 58 58 69 74 51 72 ...
 $ Gender                 : chr [1:214] "F" "M" "M" "M" ...
 $ BMI                    : chr [1:214] "NI" "NI" "25.8" "33.1" ...
 $ Syntoms onset          : chr [1:214] "44040" "44041" "NI" "44026" ...
 $ Hospitalization date   : chr [1:214] "44047" "44045" "44024" "44034" ...
 $ Admission to ICU       : chr [1:214] "No" "No" "Yes" "Yes" ...
 $ Days in ICU            : chr [1:214] "0" "0" "10" "13" ...
 $ ARDS Diagnosis         : chr [1:214] "No" "Yes" "Yes" "Yes" ...
 $ Days with ARDS         : chr [1:214] "0" "4" "10" "19" ...
 $ CRS Diagnosis          : chr [1:214] "No" "No" "Yes" "Yes" ...
 $ Days with CRS          : chr [1:214] "0" "0" "10" "19" ...
 $ Use of NIV             : chr [1:214] "No" "Yes" "No" "Yes" ...
 $ Days with NIV          : chr [1:214] "0" "5" "0" "6" ...
 $ Use of AMV             : chr [1:214] "No" "No" "Yes" "Yes" ...
 $ Days with AMV          : num [1:214] 0 0 10 13 11 5 13 0 0 6 ...
 $ Days of hospitalization: num [1:214] 8 5 10 19 12 16 26 25 6 6 ...
 $ Death                  : chr [1:214] "No" "No" "Yes" "Yes" ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1"></a><span class="fu">head</span>(clinical_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 18
  ID              Age Gender BMI   `Syntoms onset` `Hospitalization date`
  &lt;chr&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;           &lt;chr&gt;                 
1 COAG-HSJD-001    63 F      NI    44040           44047                 
2 COAG-HSJD-002    52 M      NI    44041           44045                 
3 COAG-HSJD-003    74 M      25.8  NI              44024                 
4 COAG-HSJD-004    78 M      33.1  44026           44034                 
5 COAG-HSJD-005    58 M      NI    44022           44032                 
6 COAG-HSJD-006    58 M      NI    44023           44033                 
# ℹ 12 more variables: `Admission to ICU` &lt;chr&gt;, `Days in ICU` &lt;chr&gt;,
#   `ARDS Diagnosis` &lt;chr&gt;, `Days with ARDS` &lt;chr&gt;, `CRS Diagnosis` &lt;chr&gt;,
#   `Days with CRS` &lt;chr&gt;, `Use of NIV` &lt;chr&gt;, `Days with NIV` &lt;chr&gt;,
#   `Use of AMV` &lt;chr&gt;, `Days with AMV` &lt;dbl&gt;, `Days of hospitalization` &lt;dbl&gt;,
#   Death &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1"></a><span class="co"># Convert Death column to binary (0 = No, 1 = Yes)</span></span>
<span id="cb67-2"><a href="#cb67-2"></a>clinical_data<span class="sc">$</span>Death <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(clinical_data<span class="sc">$</span>Death <span class="sc">==</span> <span class="st">"Yes"</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb67-3"><a href="#cb67-3"></a></span>
<span id="cb67-4"><a href="#cb67-4"></a><span class="co"># Convert categorical variables to factors</span></span>
<span id="cb67-5"><a href="#cb67-5"></a>clinical_data<span class="sc">$</span>Gender <span class="ot">&lt;-</span> <span class="fu">factor</span>(clinical_data<span class="sc">$</span>Gender, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"F"</span>))</span>
<span id="cb67-6"><a href="#cb67-6"></a>clinical_data<span class="sc">$</span><span class="st">`</span><span class="at">Admission to ICU</span><span class="st">`</span> <span class="ot">&lt;-</span> <span class="fu">factor</span>(clinical_data<span class="sc">$</span><span class="st">`</span><span class="at">Admission to ICU</span><span class="st">`</span>, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"No"</span>, <span class="st">"Yes"</span>))</span>
<span id="cb67-7"><a href="#cb67-7"></a>clinical_data<span class="sc">$</span><span class="st">`</span><span class="at">ARDS Diagnosis</span><span class="st">`</span> <span class="ot">&lt;-</span> <span class="fu">factor</span>(clinical_data<span class="sc">$</span><span class="st">`</span><span class="at">ARDS Diagnosis</span><span class="st">`</span>, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"No"</span>, <span class="st">"Yes"</span>))</span>
<span id="cb67-8"><a href="#cb67-8"></a></span>
<span id="cb67-9"><a href="#cb67-9"></a><span class="co"># Create Clinical Classification (G1-G4)</span></span>
<span id="cb67-10"><a href="#cb67-10"></a>clinical_data<span class="sc">$</span>Group <span class="ot">&lt;-</span> <span class="fu">with</span>(clinical_data, <span class="fu">ifelse</span>(</span>
<span id="cb67-11"><a href="#cb67-11"></a>  <span class="st">`</span><span class="at">ARDS Diagnosis</span><span class="st">`</span> <span class="sc">==</span> <span class="st">"No"</span> <span class="sc">&amp;</span> <span class="st">`</span><span class="at">Use of AMV</span><span class="st">`</span> <span class="sc">==</span> <span class="st">"No"</span> <span class="sc">&amp;</span> <span class="st">`</span><span class="at">Use of NIV</span><span class="st">`</span> <span class="sc">==</span> <span class="st">"No"</span>, <span class="st">"G1"</span>,</span>
<span id="cb67-12"><a href="#cb67-12"></a>  <span class="fu">ifelse</span>(<span class="st">`</span><span class="at">ARDS Diagnosis</span><span class="st">`</span> <span class="sc">==</span> <span class="st">"No"</span> <span class="sc">&amp;</span> <span class="st">`</span><span class="at">Use of AMV</span><span class="st">`</span> <span class="sc">!=</span> <span class="st">"No"</span>, <span class="st">"G2"</span>,</span>
<span id="cb67-13"><a href="#cb67-13"></a>         <span class="fu">ifelse</span>(<span class="st">`</span><span class="at">ARDS Diagnosis</span><span class="st">`</span> <span class="sc">==</span> <span class="st">"Yes"</span> <span class="sc">&amp;</span> <span class="st">`</span><span class="at">Use of AMV</span><span class="st">`</span> <span class="sc">==</span> <span class="st">"No"</span>, <span class="st">"G3"</span>, <span class="st">"G4"</span>)</span>
<span id="cb67-14"><a href="#cb67-14"></a>  )</span>
<span id="cb67-15"><a href="#cb67-15"></a>))</span>
<span id="cb67-16"><a href="#cb67-16"></a></span>
<span id="cb67-17"><a href="#cb67-17"></a><span class="co"># Convert Group to factor</span></span>
<span id="cb67-18"><a href="#cb67-18"></a>clinical_data<span class="sc">$</span>Group <span class="ot">&lt;-</span> <span class="fu">factor</span>(clinical_data<span class="sc">$</span>Group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"G1"</span>, <span class="st">"G2"</span>, <span class="st">"G3"</span>, <span class="st">"G4"</span>))</span>
<span id="cb67-19"><a href="#cb67-19"></a></span>
<span id="cb67-20"><a href="#cb67-20"></a><span class="co"># Load necessary library</span></span>
<span id="cb67-21"><a href="#cb67-21"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb67-22"><a href="#cb67-22"></a></span>
<span id="cb67-23"><a href="#cb67-23"></a><span class="co"># Set up layout for 4 panels (A, B, C, D)</span></span>
<span id="cb67-24"><a href="#cb67-24"></a><span class="fu">layout</span>(<span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>), <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>), <span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb67-25"><a href="#cb67-25"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">2</span>)) </span>
<span id="cb67-26"><a href="#cb67-26"></a></span>
<span id="cb67-27"><a href="#cb67-27"></a><span class="co"># ------------------- PANEL A: Age Distribution (Histogram) -------------------</span></span>
<span id="cb67-28"><a href="#cb67-28"></a><span class="fu">hist</span>(clinical_data<span class="sc">$</span>Age, </span>
<span id="cb67-29"><a href="#cb67-29"></a>     <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">20</span>, <span class="dv">100</span>, <span class="at">by =</span> <span class="dv">10</span>), </span>
<span id="cb67-30"><a href="#cb67-30"></a>     <span class="at">col =</span> <span class="st">"#e0edec"</span>, </span>
<span id="cb67-31"><a href="#cb67-31"></a>     <span class="at">border =</span> <span class="st">"black"</span>,</span>
<span id="cb67-32"><a href="#cb67-32"></a>     <span class="at">main =</span> <span class="st">"Age"</span>,</span>
<span id="cb67-33"><a href="#cb67-33"></a>     <span class="at">xlab =</span> <span class="st">"Age (years)"</span>, </span>
<span id="cb67-34"><a href="#cb67-34"></a>     <span class="at">ylab =</span> <span class="st">"Frequency (n)"</span>, </span>
<span id="cb67-35"><a href="#cb67-35"></a>     <span class="at">axes =</span> <span class="cn">TRUE</span>)</span>
<span id="cb67-36"><a href="#cb67-36"></a></span>
<span id="cb67-37"><a href="#cb67-37"></a><span class="co"># ------------------- PANEL B: Clinical Classification Table -------------------</span></span>
<span id="cb67-38"><a href="#cb67-38"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>)) <span class="co"># Remove margins</span></span>
<span id="cb67-39"><a href="#cb67-39"></a><span class="fu">plot.new</span>()  <span class="co"># Create blank plot area</span></span>
<span id="cb67-40"><a href="#cb67-40"></a></span>
<span id="cb67-41"><a href="#cb67-41"></a><span class="co"># Add small title</span></span>
<span id="cb67-42"><a href="#cb67-42"></a><span class="fu">title</span>(<span class="at">xlab =</span> <span class="fu">expression</span>(<span class="fu">bold</span>(<span class="st">"Definition of the clinical classification"</span>)), </span>
<span id="cb67-43"><a href="#cb67-43"></a>      <span class="at">line =</span> <span class="sc">-</span><span class="fl">12.5</span>, <span class="at">adj =</span> <span class="fl">0.5</span>, <span class="at">cex.lab =</span> <span class="fl">1.2</span>) </span>
<span id="cb67-44"><a href="#cb67-44"></a></span>
<span id="cb67-45"><a href="#cb67-45"></a><span class="co"># Define table content</span></span>
<span id="cb67-46"><a href="#cb67-46"></a>table_content <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">data.frame</span>(</span>
<span id="cb67-47"><a href="#cb67-47"></a>  <span class="st">"Clinical</span><span class="sc">\n</span><span class="st">Classification"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"G1"</span>, <span class="st">"G2"</span>, <span class="st">"G3"</span>, <span class="st">"G4"</span>),</span>
<span id="cb67-48"><a href="#cb67-48"></a>  <span class="at">NIV =</span> <span class="fu">c</span>(<span class="st">"−"</span>, <span class="st">"−/+"</span>, <span class="st">"+"</span>, <span class="st">"−/+"</span>),</span>
<span id="cb67-49"><a href="#cb67-49"></a>  <span class="at">AMV =</span> <span class="fu">c</span>(<span class="st">"−"</span>, <span class="st">"+/−"</span>, <span class="st">"−"</span>, <span class="st">"+"</span>),</span>
<span id="cb67-50"><a href="#cb67-50"></a>  <span class="at">ARDS =</span> <span class="fu">c</span>(<span class="st">"−"</span>, <span class="st">"−"</span>, <span class="st">"+"</span>, <span class="st">"+"</span>),</span>
<span id="cb67-51"><a href="#cb67-51"></a>  <span class="at">check.names =</span> <span class="cn">FALSE</span>  <span class="co"># Prevents automatic conversion of spaces to underscores</span></span>
<span id="cb67-52"><a href="#cb67-52"></a>))</span>
<span id="cb67-53"><a href="#cb67-53"></a></span>
<span id="cb67-54"><a href="#cb67-54"></a><span class="co"># Define table styling</span></span>
<span id="cb67-55"><a href="#cb67-55"></a>table_theme <span class="ot">&lt;-</span> <span class="fu">ttheme_default</span>(</span>
<span id="cb67-56"><a href="#cb67-56"></a>  <span class="at">colhead =</span> <span class="fu">list</span>(</span>
<span id="cb67-57"><a href="#cb67-57"></a>    <span class="at">bg_params =</span> <span class="fu">list</span>(<span class="at">fill =</span> <span class="st">"#d1d0d0"</span>),  <span class="co"># Darker gray for column headers</span></span>
<span id="cb67-58"><a href="#cb67-58"></a>    <span class="at">fg_params =</span> <span class="fu">list</span>(<span class="at">fontface =</span> <span class="st">"bold"</span>)  <span class="co"># Bold for column headers</span></span>
<span id="cb67-59"><a href="#cb67-59"></a>  ),</span>
<span id="cb67-60"><a href="#cb67-60"></a>   <span class="at">colhead =</span> <span class="fu">list</span>(</span>
<span id="cb67-61"><a href="#cb67-61"></a>    <span class="at">fg_params =</span> <span class="fu">list</span>(<span class="at">fontface =</span> <span class="st">"bold"</span>, <span class="at">fontsize =</span> <span class="dv">6</span>)  <span class="co"># Smaller and bold font for column headers</span></span>
<span id="cb67-62"><a href="#cb67-62"></a>  )</span>
<span id="cb67-63"><a href="#cb67-63"></a>)</span>
<span id="cb67-64"><a href="#cb67-64"></a></span>
<span id="cb67-65"><a href="#cb67-65"></a><span class="co"># Force table into upper-right panel</span></span>
<span id="cb67-66"><a href="#cb67-66"></a>vp <span class="ot">&lt;-</span> <span class="fu">viewport</span>(<span class="at">x =</span> <span class="fl">0.52</span>, <span class="at">y =</span> <span class="fl">1.1</span>, <span class="at">width =</span> <span class="fl">0.5</span>, <span class="at">height =</span> <span class="fl">0.8</span>, <span class="at">just =</span> <span class="fu">c</span>(<span class="st">"left"</span>, <span class="st">"top"</span>))</span>
<span id="cb67-67"><a href="#cb67-67"></a><span class="fu">pushViewport</span>(vp)</span>
<span id="cb67-68"><a href="#cb67-68"></a><span class="fu">grid.table</span>(table_content, <span class="at">theme =</span> table_theme)</span>
<span id="cb67-69"><a href="#cb67-69"></a><span class="fu">popViewport</span>()</span>
<span id="cb67-70"><a href="#cb67-70"></a></span>
<span id="cb67-71"><a href="#cb67-71"></a><span class="co"># ------------------- PANEL C: Clinical Classification Barplot -------------------</span></span>
<span id="cb67-72"><a href="#cb67-72"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">2</span>)) </span>
<span id="cb67-73"><a href="#cb67-73"></a>group_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(clinical_data<span class="sc">$</span>Group)</span>
<span id="cb67-74"><a href="#cb67-74"></a>bar_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#96d0cb"</span>, <span class="st">"#f7f5b5"</span>, <span class="st">"#b9b9dc"</span>, <span class="st">"#e98f80"</span>)</span>
<span id="cb67-75"><a href="#cb67-75"></a></span>
<span id="cb67-76"><a href="#cb67-76"></a>bp <span class="ot">&lt;-</span> <span class="fu">barplot</span>(group_counts, </span>
<span id="cb67-77"><a href="#cb67-77"></a>              <span class="at">col =</span> bar_colors, </span>
<span id="cb67-78"><a href="#cb67-78"></a>              <span class="at">border =</span> <span class="st">"black"</span>, </span>
<span id="cb67-79"><a href="#cb67-79"></a>              <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">80</span>), </span>
<span id="cb67-80"><a href="#cb67-80"></a>              <span class="at">main =</span> <span class="st">"Clinical classification"</span>, </span>
<span id="cb67-81"><a href="#cb67-81"></a>              <span class="at">ylab =</span> <span class="st">"Frequency (n)"</span>,</span>
<span id="cb67-82"><a href="#cb67-82"></a>              <span class="at">xlab =</span> <span class="st">"Clinical classification"</span>,</span>
<span id="cb67-83"><a href="#cb67-83"></a>              <span class="at">names.arg =</span> <span class="fu">c</span>(<span class="st">"G1"</span>, <span class="st">"G2"</span>, <span class="st">"G3"</span>, <span class="st">"G4"</span>))</span>
<span id="cb67-84"><a href="#cb67-84"></a></span>
<span id="cb67-85"><a href="#cb67-85"></a><span class="co"># Add numbers above bars</span></span>
<span id="cb67-86"><a href="#cb67-86"></a><span class="fu">text</span>(bp, group_counts <span class="sc">+</span> <span class="dv">3</span>, <span class="at">labels =</span> group_counts, <span class="at">cex =</span> <span class="dv">1</span>)</span>
<span id="cb67-87"><a href="#cb67-87"></a></span>
<span id="cb67-88"><a href="#cb67-88"></a><span class="co"># ------------------- PANEL D: Vital Status Barplot -------------------</span></span>
<span id="cb67-89"><a href="#cb67-89"></a>death_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(clinical_data<span class="sc">$</span>Death)</span>
<span id="cb67-90"><a href="#cb67-90"></a>bp <span class="ot">&lt;-</span> <span class="fu">barplot</span>(death_counts, </span>
<span id="cb67-91"><a href="#cb67-91"></a>              <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#96d0cb"</span>, <span class="st">"#f7f5b5"</span>), </span>
<span id="cb67-92"><a href="#cb67-92"></a>              <span class="at">border =</span> <span class="st">"black"</span>, </span>
<span id="cb67-93"><a href="#cb67-93"></a>              <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">160</span>), </span>
<span id="cb67-94"><a href="#cb67-94"></a>              <span class="at">main =</span> <span class="st">"Vital status"</span>, </span>
<span id="cb67-95"><a href="#cb67-95"></a>              <span class="at">ylab =</span> <span class="st">"Frequency (n)"</span>,</span>
<span id="cb67-96"><a href="#cb67-96"></a>              <span class="at">xlab=</span><span class="st">"Death"</span>,</span>
<span id="cb67-97"><a href="#cb67-97"></a>              <span class="at">names.arg =</span> <span class="fu">c</span>(<span class="st">"No"</span>, <span class="st">"Yes"</span>))</span>
<span id="cb67-98"><a href="#cb67-98"></a></span>
<span id="cb67-99"><a href="#cb67-99"></a><span class="co"># Add numbers directly above bars</span></span>
<span id="cb67-100"><a href="#cb67-100"></a><span class="fu">text</span>(bp, death_counts <span class="sc">+</span> <span class="dv">5</span>, <span class="at">labels =</span> death_counts, <span class="at">cex =</span> <span class="dv">1</span>)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>I tried my best, but I really don’t know how to center the table and make it look correct because it has a fixed size while the plots adjust wheher n we move in and out. I think the easiest approach would be to create the separate plots and the table and merge them with another tool.</p>
</section>
<section id="reproduce-figure-2-from-the-publication-but-instead-of-representing-the-clusters-in-the-annotation-represent-the-groups-g1-to-g4" class="level2">
<h2 class="anchored" data-anchor-id="reproduce-figure-2-from-the-publication-but-instead-of-representing-the-clusters-in-the-annotation-represent-the-groups-g1-to-g4">2.2. Reproduce Figure 2 from the publication but instead of representing the clusters in the annotation, represent the groups (G1 to G4)</h2>
<section id="loading-and-cleaning-the-data" class="level3">
<h3 class="anchored" data-anchor-id="loading-and-cleaning-the-data">2.2.1. Loading and cleaning the data</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1"></a><span class="fu">library</span>(circlize)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>========================================
circlize version 0.4.16
CRAN page: https://cran.r-project.org/package=circlize
Github page: https://github.com/jokergoo/circlize
Documentation: https://jokergoo.github.io/circlize_book/book/

If you use it in published research, please cite:
Gu, Z. circlize implements and enhances circular visualization
  in R. Bioinformatics 2014.

This message can be suppressed by:
  suppressPackageStartupMessages(library(circlize))
========================================</code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1"></a><span class="co"># Load the data for cytokines</span></span>
<span id="cb70-2"><a href="#cb70-2"></a>clinical_data_2 <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"data_article/Table 2.XLSX"</span>, <span class="at">skip =</span> <span class="dv">1</span>)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>New names:
• `` -&gt; `...1`</code></pre>
</div>
</div>
<p>If we take a look at the data we will see that we have many missing values on the first columns (ID), that is because if we have more than one measure, taken in different days, only the first measure has the ID. So we want fill the missing IDs with the last non-missing value. We can do so with the function <code>fill()</code> from the package <code>tidyr</code>.</p>
<p>Afterwards, the investigators performed the clustering using the data from 13 cytokines: IL-1 beta, IL-6, IL-8, IL-10, IL-17, TNF, IFNalpha, IFN-gamma, CXCL10, CCL2, CCL3, G-CSF and IL-38. However on the data-set and the image IL-17 is missing.</p>
<p>Additionally, we have missing values, marked as NI, so we want to change them to NA.</p>
<p>Finally, now some patients have many cytokines values, which we don’t want. According to the article, <em>only the median value of each cytokine or laboratory parameter was considered to compare patients</em>. So we will do the same. To do so we will transform our data-set to a long format, using <code>pivot_longer()</code>, which will contains the IDs and Day as normal and the cytokine name. Then we will make sure all the measurements are numeric and finally we will perform the mean.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1"></a><span class="fu">colnames</span>(clinical_data_2)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"ID"</span></span>
<span id="cb72-2"><a href="#cb72-2"></a>clinical_data_2 <span class="ot">&lt;-</span> clinical_data_2 <span class="sc">|&gt;</span></span>
<span id="cb72-3"><a href="#cb72-3"></a>  <span class="fu">fill</span>(ID, <span class="at">.direction =</span> <span class="st">"down"</span>)</span>
<span id="cb72-4"><a href="#cb72-4"></a></span>
<span id="cb72-5"><a href="#cb72-5"></a>clinical_data_2[clinical_data_2 <span class="sc">==</span> <span class="st">"NI"</span>] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb72-6"><a href="#cb72-6"></a></span>
<span id="cb72-7"><a href="#cb72-7"></a>clinical_data_2 <span class="ot">&lt;-</span> clinical_data_2[, <span class="dv">1</span><span class="sc">:</span><span class="dv">14</span>]</span>
<span id="cb72-8"><a href="#cb72-8"></a></span>
<span id="cb72-9"><a href="#cb72-9"></a>long_clinical_data_2 <span class="ot">&lt;-</span> clinical_data_2 <span class="sc">|&gt;</span></span>
<span id="cb72-10"><a href="#cb72-10"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb72-11"><a href="#cb72-11"></a>    <span class="at">cols =</span> <span class="fu">c</span>(<span class="sc">-</span>ID, <span class="sc">-</span>Day),                        </span>
<span id="cb72-12"><a href="#cb72-12"></a>    <span class="at">names_to =</span> <span class="st">"Cytokine"</span>,           </span>
<span id="cb72-13"><a href="#cb72-13"></a>    <span class="at">values_to =</span> <span class="st">"Measurement"</span></span>
<span id="cb72-14"><a href="#cb72-14"></a>  )</span>
<span id="cb72-15"><a href="#cb72-15"></a>  </span>
<span id="cb72-16"><a href="#cb72-16"></a>long_clinical_data_2<span class="sc">$</span>Measurement <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(long_clinical_data_2<span class="sc">$</span>Measurement)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: NAs introduced by coercion</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1"></a>average_clinical_data_2 <span class="ot">&lt;-</span> long_clinical_data_2 <span class="sc">|&gt;</span></span>
<span id="cb74-2"><a href="#cb74-2"></a>  <span class="fu">group_by</span>(ID, Cytokine) <span class="sc">|&gt;</span></span>
<span id="cb74-3"><a href="#cb74-3"></a>  <span class="fu">summarise</span>(<span class="at">Mean_Measurement =</span> <span class="fu">mean</span>(Measurement, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.groups =</span> <span class="st">"drop"</span>)</span>
<span id="cb74-4"><a href="#cb74-4"></a></span>
<span id="cb74-5"><a href="#cb74-5"></a>final_clinical_data_2 <span class="ot">&lt;-</span> average_clinical_data_2 <span class="sc">|&gt;</span></span>
<span id="cb74-6"><a href="#cb74-6"></a>  <span class="fu">pivot_wider</span>(</span>
<span id="cb74-7"><a href="#cb74-7"></a>    <span class="at">names_from =</span> Cytokine,          </span>
<span id="cb74-8"><a href="#cb74-8"></a>    <span class="at">values_from =</span> Mean_Measurement    </span>
<span id="cb74-9"><a href="#cb74-9"></a>  )</span>
<span id="cb74-10"><a href="#cb74-10"></a></span>
<span id="cb74-11"><a href="#cb74-11"></a>final_clinical_data_2 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(final_clinical_data_2)</span>
<span id="cb74-12"><a href="#cb74-12"></a></span>
<span id="cb74-13"><a href="#cb74-13"></a><span class="fu">dim</span>(final_clinical_data_2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 194  13</code></pre>
</div>
</div>
<p>Now we have a data-set that is almost like the one the investigators used, with 194 patients as the rows and 12 cytokines as columns instead of 13.</p>
<p>It is not mentioned if some sort of standardization was performed, on Figure 2 it says that <em>median value for each cytokine was used to create profiles or clusters</em>. Thus, we consider that we can use the values as they are currently.</p>
<p>However, there are some patients that lack IL-6 values, to avoid losing information, we will try to generate the data using the function <code>mice()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1"></a>imputed_data <span class="ot">&lt;-</span> <span class="fu">mice</span>(final_clinical_data_2, <span class="at">m =</span> <span class="dv">5</span>, <span class="at">method =</span> <span class="st">'pmm'</span>, <span class="at">seed =</span> <span class="dv">1</span>, <span class="at">maxit =</span> <span class="dv">100</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 iter imp variable
  1   1  IL-6
  1   2  IL-6
  1   3  IL-6
  1   4  IL-6
  1   5  IL-6
  2   1  IL-6
  2   2  IL-6
  2   3  IL-6
  2   4  IL-6
  2   5  IL-6
  3   1  IL-6
  3   2  IL-6
  3   3  IL-6
  3   4  IL-6
  3   5  IL-6
  4   1  IL-6
  4   2  IL-6
  4   3  IL-6
  4   4  IL-6
  4   5  IL-6
  5   1  IL-6
  5   2  IL-6
  5   3  IL-6
  5   4  IL-6
  5   5  IL-6
  6   1  IL-6
  6   2  IL-6
  6   3  IL-6
  6   4  IL-6
  6   5  IL-6
  7   1  IL-6
  7   2  IL-6
  7   3  IL-6
  7   4  IL-6
  7   5  IL-6
  8   1  IL-6
  8   2  IL-6
  8   3  IL-6
  8   4  IL-6
  8   5  IL-6
  9   1  IL-6
  9   2  IL-6
  9   3  IL-6
  9   4  IL-6
  9   5  IL-6
  10   1  IL-6
  10   2  IL-6
  10   3  IL-6
  10   4  IL-6
  10   5  IL-6
  11   1  IL-6
  11   2  IL-6
  11   3  IL-6
  11   4  IL-6
  11   5  IL-6
  12   1  IL-6
  12   2  IL-6
  12   3  IL-6
  12   4  IL-6
  12   5  IL-6
  13   1  IL-6
  13   2  IL-6
  13   3  IL-6
  13   4  IL-6
  13   5  IL-6
  14   1  IL-6
  14   2  IL-6
  14   3  IL-6
  14   4  IL-6
  14   5  IL-6
  15   1  IL-6
  15   2  IL-6
  15   3  IL-6
  15   4  IL-6
  15   5  IL-6
  16   1  IL-6
  16   2  IL-6
  16   3  IL-6
  16   4  IL-6
  16   5  IL-6
  17   1  IL-6
  17   2  IL-6
  17   3  IL-6
  17   4  IL-6
  17   5  IL-6
  18   1  IL-6
  18   2  IL-6
  18   3  IL-6
  18   4  IL-6
  18   5  IL-6
  19   1  IL-6
  19   2  IL-6
  19   3  IL-6
  19   4  IL-6
  19   5  IL-6
  20   1  IL-6
  20   2  IL-6
  20   3  IL-6
  20   4  IL-6
  20   5  IL-6
  21   1  IL-6
  21   2  IL-6
  21   3  IL-6
  21   4  IL-6
  21   5  IL-6
  22   1  IL-6
  22   2  IL-6
  22   3  IL-6
  22   4  IL-6
  22   5  IL-6
  23   1  IL-6
  23   2  IL-6
  23   3  IL-6
  23   4  IL-6
  23   5  IL-6
  24   1  IL-6
  24   2  IL-6
  24   3  IL-6
  24   4  IL-6
  24   5  IL-6
  25   1  IL-6
  25   2  IL-6
  25   3  IL-6
  25   4  IL-6
  25   5  IL-6
  26   1  IL-6
  26   2  IL-6
  26   3  IL-6
  26   4  IL-6
  26   5  IL-6
  27   1  IL-6
  27   2  IL-6
  27   3  IL-6
  27   4  IL-6
  27   5  IL-6
  28   1  IL-6
  28   2  IL-6
  28   3  IL-6
  28   4  IL-6
  28   5  IL-6
  29   1  IL-6
  29   2  IL-6
  29   3  IL-6
  29   4  IL-6
  29   5  IL-6
  30   1  IL-6
  30   2  IL-6
  30   3  IL-6
  30   4  IL-6
  30   5  IL-6
  31   1  IL-6
  31   2  IL-6
  31   3  IL-6
  31   4  IL-6
  31   5  IL-6
  32   1  IL-6
  32   2  IL-6
  32   3  IL-6
  32   4  IL-6
  32   5  IL-6
  33   1  IL-6
  33   2  IL-6
  33   3  IL-6
  33   4  IL-6
  33   5  IL-6
  34   1  IL-6
  34   2  IL-6
  34   3  IL-6
  34   4  IL-6
  34   5  IL-6
  35   1  IL-6
  35   2  IL-6
  35   3  IL-6
  35   4  IL-6
  35   5  IL-6
  36   1  IL-6
  36   2  IL-6
  36   3  IL-6
  36   4  IL-6
  36   5  IL-6
  37   1  IL-6
  37   2  IL-6
  37   3  IL-6
  37   4  IL-6
  37   5  IL-6
  38   1  IL-6
  38   2  IL-6
  38   3  IL-6
  38   4  IL-6
  38   5  IL-6
  39   1  IL-6
  39   2  IL-6
  39   3  IL-6
  39   4  IL-6
  39   5  IL-6
  40   1  IL-6
  40   2  IL-6
  40   3  IL-6
  40   4  IL-6
  40   5  IL-6
  41   1  IL-6
  41   2  IL-6
  41   3  IL-6
  41   4  IL-6
  41   5  IL-6
  42   1  IL-6
  42   2  IL-6
  42   3  IL-6
  42   4  IL-6
  42   5  IL-6
  43   1  IL-6
  43   2  IL-6
  43   3  IL-6
  43   4  IL-6
  43   5  IL-6
  44   1  IL-6
  44   2  IL-6
  44   3  IL-6
  44   4  IL-6
  44   5  IL-6
  45   1  IL-6
  45   2  IL-6
  45   3  IL-6
  45   4  IL-6
  45   5  IL-6
  46   1  IL-6
  46   2  IL-6
  46   3  IL-6
  46   4  IL-6
  46   5  IL-6
  47   1  IL-6
  47   2  IL-6
  47   3  IL-6
  47   4  IL-6
  47   5  IL-6
  48   1  IL-6
  48   2  IL-6
  48   3  IL-6
  48   4  IL-6
  48   5  IL-6
  49   1  IL-6
  49   2  IL-6
  49   3  IL-6
  49   4  IL-6
  49   5  IL-6
  50   1  IL-6
  50   2  IL-6
  50   3  IL-6
  50   4  IL-6
  50   5  IL-6
  51   1  IL-6
  51   2  IL-6
  51   3  IL-6
  51   4  IL-6
  51   5  IL-6
  52   1  IL-6
  52   2  IL-6
  52   3  IL-6
  52   4  IL-6
  52   5  IL-6
  53   1  IL-6
  53   2  IL-6
  53   3  IL-6
  53   4  IL-6
  53   5  IL-6
  54   1  IL-6
  54   2  IL-6
  54   3  IL-6
  54   4  IL-6
  54   5  IL-6
  55   1  IL-6
  55   2  IL-6
  55   3  IL-6
  55   4  IL-6
  55   5  IL-6
  56   1  IL-6
  56   2  IL-6
  56   3  IL-6
  56   4  IL-6
  56   5  IL-6
  57   1  IL-6
  57   2  IL-6
  57   3  IL-6
  57   4  IL-6
  57   5  IL-6
  58   1  IL-6
  58   2  IL-6
  58   3  IL-6
  58   4  IL-6
  58   5  IL-6
  59   1  IL-6
  59   2  IL-6
  59   3  IL-6
  59   4  IL-6
  59   5  IL-6
  60   1  IL-6
  60   2  IL-6
  60   3  IL-6
  60   4  IL-6
  60   5  IL-6
  61   1  IL-6
  61   2  IL-6
  61   3  IL-6
  61   4  IL-6
  61   5  IL-6
  62   1  IL-6
  62   2  IL-6
  62   3  IL-6
  62   4  IL-6
  62   5  IL-6
  63   1  IL-6
  63   2  IL-6
  63   3  IL-6
  63   4  IL-6
  63   5  IL-6
  64   1  IL-6
  64   2  IL-6
  64   3  IL-6
  64   4  IL-6
  64   5  IL-6
  65   1  IL-6
  65   2  IL-6
  65   3  IL-6
  65   4  IL-6
  65   5  IL-6
  66   1  IL-6
  66   2  IL-6
  66   3  IL-6
  66   4  IL-6
  66   5  IL-6
  67   1  IL-6
  67   2  IL-6
  67   3  IL-6
  67   4  IL-6
  67   5  IL-6
  68   1  IL-6
  68   2  IL-6
  68   3  IL-6
  68   4  IL-6
  68   5  IL-6
  69   1  IL-6
  69   2  IL-6
  69   3  IL-6
  69   4  IL-6
  69   5  IL-6
  70   1  IL-6
  70   2  IL-6
  70   3  IL-6
  70   4  IL-6
  70   5  IL-6
  71   1  IL-6
  71   2  IL-6
  71   3  IL-6
  71   4  IL-6
  71   5  IL-6
  72   1  IL-6
  72   2  IL-6
  72   3  IL-6
  72   4  IL-6
  72   5  IL-6
  73   1  IL-6
  73   2  IL-6
  73   3  IL-6
  73   4  IL-6
  73   5  IL-6
  74   1  IL-6
  74   2  IL-6
  74   3  IL-6
  74   4  IL-6
  74   5  IL-6
  75   1  IL-6
  75   2  IL-6
  75   3  IL-6
  75   4  IL-6
  75   5  IL-6
  76   1  IL-6
  76   2  IL-6
  76   3  IL-6
  76   4  IL-6
  76   5  IL-6
  77   1  IL-6
  77   2  IL-6
  77   3  IL-6
  77   4  IL-6
  77   5  IL-6
  78   1  IL-6
  78   2  IL-6
  78   3  IL-6
  78   4  IL-6
  78   5  IL-6
  79   1  IL-6
  79   2  IL-6
  79   3  IL-6
  79   4  IL-6
  79   5  IL-6
  80   1  IL-6
  80   2  IL-6
  80   3  IL-6
  80   4  IL-6
  80   5  IL-6
  81   1  IL-6
  81   2  IL-6
  81   3  IL-6
  81   4  IL-6
  81   5  IL-6
  82   1  IL-6
  82   2  IL-6
  82   3  IL-6
  82   4  IL-6
  82   5  IL-6
  83   1  IL-6
  83   2  IL-6
  83   3  IL-6
  83   4  IL-6
  83   5  IL-6
  84   1  IL-6
  84   2  IL-6
  84   3  IL-6
  84   4  IL-6
  84   5  IL-6
  85   1  IL-6
  85   2  IL-6
  85   3  IL-6
  85   4  IL-6
  85   5  IL-6
  86   1  IL-6
  86   2  IL-6
  86   3  IL-6
  86   4  IL-6
  86   5  IL-6
  87   1  IL-6
  87   2  IL-6
  87   3  IL-6
  87   4  IL-6
  87   5  IL-6
  88   1  IL-6
  88   2  IL-6
  88   3  IL-6
  88   4  IL-6
  88   5  IL-6
  89   1  IL-6
  89   2  IL-6
  89   3  IL-6
  89   4  IL-6
  89   5  IL-6
  90   1  IL-6
  90   2  IL-6
  90   3  IL-6
  90   4  IL-6
  90   5  IL-6
  91   1  IL-6
  91   2  IL-6
  91   3  IL-6
  91   4  IL-6
  91   5  IL-6
  92   1  IL-6
  92   2  IL-6
  92   3  IL-6
  92   4  IL-6
  92   5  IL-6
  93   1  IL-6
  93   2  IL-6
  93   3  IL-6
  93   4  IL-6
  93   5  IL-6
  94   1  IL-6
  94   2  IL-6
  94   3  IL-6
  94   4  IL-6
  94   5  IL-6
  95   1  IL-6
  95   2  IL-6
  95   3  IL-6
  95   4  IL-6
  95   5  IL-6
  96   1  IL-6
  96   2  IL-6
  96   3  IL-6
  96   4  IL-6
  96   5  IL-6
  97   1  IL-6
  97   2  IL-6
  97   3  IL-6
  97   4  IL-6
  97   5  IL-6
  98   1  IL-6
  98   2  IL-6
  98   3  IL-6
  98   4  IL-6
  98   5  IL-6
  99   1  IL-6
  99   2  IL-6
  99   3  IL-6
  99   4  IL-6
  99   5  IL-6
  100   1  IL-6
  100   2  IL-6
  100   3  IL-6
  100   4  IL-6
  100   5  IL-6</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Number of logged events: 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1"></a>final_clinical_data_2 <span class="ot">&lt;-</span> <span class="fu">complete</span>(imputed_data, <span class="dv">1</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, we want to add the group of each patient.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1"></a>labels <span class="ot">&lt;-</span> clinical_data[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">19</span>)]</span>
<span id="cb80-2"><a href="#cb80-2"></a>labels <span class="ot">&lt;-</span> labels[<span class="fu">which</span>(<span class="fu">duplicated</span>(labels<span class="sc">$</span>ID)<span class="sc">==</span><span class="cn">FALSE</span>),]</span>
<span id="cb80-3"><a href="#cb80-3"></a>labels <span class="ot">&lt;-</span> labels[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(labels<span class="sc">$</span>Group)),]</span>
<span id="cb80-4"><a href="#cb80-4"></a>final_clinical_data_2 <span class="ot">&lt;-</span> <span class="fu">merge</span>(final_clinical_data_2, labels, <span class="at">by =</span> <span class="st">"ID"</span>)</span>
<span id="cb80-5"><a href="#cb80-5"></a><span class="fu">dim</span>(final_clinical_data_2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 188  14</code></pre>
</div>
</div>
<p>Sadly we have lost some patients that don’t have an associated group.</p>
</section>
<section id="heat-map" class="level3">
<h3 class="anchored" data-anchor-id="heat-map">2.2.3. Heat map</h3>
<p>Currently on our data-set we have the raw expression of the cytokines. However, on the Figure 2 the values are in “Relative Expression”. That means we will normalize the expression and convert the minimum value to 0%, the maximum to 100% and the rest an in-between.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1"></a>normalize_to_percent <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb82-2"><a href="#cb82-2"></a>  <span class="fu">return</span>((x <span class="sc">-</span> <span class="fu">min</span>(x)) <span class="sc">/</span> (<span class="fu">max</span>(x) <span class="sc">-</span> <span class="fu">min</span>(x)) <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb82-3"><a href="#cb82-3"></a>}</span>
<span id="cb82-4"><a href="#cb82-4"></a>numeric_data <span class="ot">&lt;-</span> final_clinical_data_2[<span class="fu">sapply</span>(final_clinical_data_2, is.numeric)]</span>
<span id="cb82-5"><a href="#cb82-5"></a>non_numeric_data <span class="ot">&lt;-</span> final_clinical_data_2[, <span class="sc">!</span><span class="fu">sapply</span>(final_clinical_data_2, is.numeric)]</span>
<span id="cb82-6"><a href="#cb82-6"></a></span>
<span id="cb82-7"><a href="#cb82-7"></a>normalized_clinical_data_2 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(numeric_data, normalize_to_percent))</span>
<span id="cb82-8"><a href="#cb82-8"></a></span>
<span id="cb82-9"><a href="#cb82-9"></a>final_data_2 <span class="ot">&lt;-</span> <span class="fu">cbind</span>(non_numeric_data, normalized_clinical_data_2)</span>
<span id="cb82-10"><a href="#cb82-10"></a><span class="fu">colnames</span>(final_data_2)[<span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="st">"Group"</span></span>
<span id="cb82-11"><a href="#cb82-11"></a><span class="fu">rownames</span>(final_data_2) <span class="ot">&lt;-</span> final_data_2[,<span class="dv">1</span>]</span>
<span id="cb82-12"><a href="#cb82-12"></a>final_data_2 <span class="ot">&lt;-</span> final_data_2[, <span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb82-13"><a href="#cb82-13"></a>final_data_2<span class="sc">$</span>Group <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(final_data_2<span class="sc">$</span>Group)</span>
<span id="cb82-14"><a href="#cb82-14"></a></span>
<span id="cb82-15"><a href="#cb82-15"></a><span class="fu">summary</span>(final_data_2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Group        CCL2              CCL3              CXCL10      
 G1:56   Min.   :  0.000   Min.   :  0.0000   Min.   :  0.00  
 G2: 7   1st Qu.:  1.703   1st Qu.:  0.9824   1st Qu.: 15.93  
 G3:64   Median :  4.405   Median :  1.5266   Median : 38.84  
 G4:61   Mean   :  8.482   Mean   :  2.9592   Mean   : 42.49  
         3rd Qu.:  9.675   3rd Qu.:  2.4748   3rd Qu.: 69.72  
         Max.   :100.000   Max.   :100.0000   Max.   :100.00  
     G.CSF              IFN.ɑ              IFN.γ              IL.10         
 Min.   :  0.0000   Min.   :  0.0000   Min.   :  0.0000   Min.   :  0.0000  
 1st Qu.:  0.4112   1st Qu.:  0.1968   1st Qu.:  0.2108   1st Qu.:  0.2283  
 Median :  0.6564   Median :  0.4973   Median :  0.2522   Median :  0.4458  
 Mean   :  2.4070   Mean   :  3.4297   Mean   :  1.1756   Mean   :  1.8167  
 3rd Qu.:  1.4298   3rd Qu.:  2.2862   3rd Qu.:  0.3512   3rd Qu.:  0.8265  
 Max.   :100.0000   Max.   :100.0000   Max.   :100.0000   Max.   :100.0000  
     IL.1β             IL.38             IL.6                IL.8        
 Min.   :  0.000   Min.   :  0.00   Min.   :  0.00000   Min.   :  0.000  
 1st Qu.:  0.000   1st Qu.: 16.02   1st Qu.:  0.01646   1st Qu.:  2.259  
 Median :  0.000   Median : 27.42   Median :  0.08268   Median :  4.040  
 Mean   :  2.276   Mean   : 32.12   Mean   :  7.71814   Mean   :  8.335  
 3rd Qu.:  0.000   3rd Qu.: 45.22   3rd Qu.:  0.33139   3rd Qu.:  7.817  
 Max.   :100.000   Max.   :100.00   Max.   :100.00000   Max.   :100.000  
     TNF.ɑ          
 Min.   :  0.00000  
 1st Qu.:  0.08302  
 Median :  0.14395  
 Mean   :  2.33178  
 3rd Qu.:  0.42365  
 Max.   :100.00000  </code></pre>
</div>
</div>
<p>We can see that indeed all numerical attributes have a minimum of 0 and a maximum of 100. Now we can plot the heat-map.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb84-2"><a href="#cb84-2"></a></span>
<span id="cb84-3"><a href="#cb84-3"></a>final_data_2_heatmap <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(final_data_2[, <span class="sc">-</span><span class="fu">which</span>(<span class="fu">names</span>(final_data_2) <span class="sc">==</span> <span class="st">"Group"</span>)])</span>
<span id="cb84-4"><a href="#cb84-4"></a></span>
<span id="cb84-5"><a href="#cb84-5"></a>cytokine_order <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"IL.6"</span>, <span class="st">"CXCL10"</span>, <span class="st">"IL.38"</span>, <span class="st">"IL.8"</span>, <span class="st">"IFN.ɑ"</span>, <span class="st">"IL.10"</span>, <span class="st">"TNF.ɑ"</span>, <span class="st">"CCL2"</span>, <span class="st">"CCL3"</span>, <span class="st">"IFN.γ"</span>, <span class="st">"IL.1β"</span>, <span class="st">"G.CSF"</span>)</span>
<span id="cb84-6"><a href="#cb84-6"></a></span>
<span id="cb84-7"><a href="#cb84-7"></a>final_data_2_heatmap <span class="ot">&lt;-</span> final_data_2_heatmap[, cytokine_order]</span>
<span id="cb84-8"><a href="#cb84-8"></a></span>
<span id="cb84-9"><a href="#cb84-9"></a>orange_palette <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">9</span>, <span class="st">"Oranges"</span>)</span>
<span id="cb84-10"><a href="#cb84-10"></a>group_colors <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">4</span>, <span class="st">"Set2"</span>) </span>
<span id="cb84-11"><a href="#cb84-11"></a></span>
<span id="cb84-12"><a href="#cb84-12"></a><span class="fu">pheatmap</span>(<span class="fu">t</span>(final_data_2_heatmap),</span>
<span id="cb84-13"><a href="#cb84-13"></a>         <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb84-14"><a href="#cb84-14"></a>         <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>,</span>
<span id="cb84-15"><a href="#cb84-15"></a>         <span class="at">show_rownames =</span> <span class="cn">TRUE</span>, </span>
<span id="cb84-16"><a href="#cb84-16"></a>         <span class="at">show_colnames =</span> <span class="cn">FALSE</span>, </span>
<span id="cb84-17"><a href="#cb84-17"></a>         <span class="at">color =</span> orange_palette,  </span>
<span id="cb84-18"><a href="#cb84-18"></a>         <span class="at">annotation_col =</span> non_numeric_data[, <span class="st">"Group"</span>, <span class="at">drop =</span> <span class="cn">FALSE</span>],  </span>
<span id="cb84-19"><a href="#cb84-19"></a>        <span class="at">annotation_colors =</span> <span class="fu">list</span>(<span class="at">Group =</span> <span class="fu">setNames</span>(group_colors, <span class="fu">c</span>(<span class="st">"G1"</span>, <span class="st">"G2"</span>, <span class="st">"G3"</span>, <span class="st">"G4"</span>))),</span>
<span id="cb84-20"><a href="#cb84-20"></a>         <span class="at">legend =</span> <span class="cn">TRUE</span>,</span>
<span id="cb84-21"><a href="#cb84-21"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="improve-figure-2-of-the-publication" class="level2">
<h2 class="anchored" data-anchor-id="improve-figure-2-of-the-publication">2.3. Improve Figure 2 of the publication</h2>
<p>Add a second annotation with information of death and a third one with information of gender</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1"></a>death <span class="ot">&lt;-</span> clinical_data[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">18</span>)]</span>
<span id="cb85-2"><a href="#cb85-2"></a><span class="co"># gender &lt;- clinical_data[, c(1, 3)]</span></span>
<span id="cb85-3"><a href="#cb85-3"></a>final_data_2<span class="sc">$</span>ID <span class="ot">&lt;-</span> <span class="fu">rownames</span>(final_data_2)</span>
<span id="cb85-4"><a href="#cb85-4"></a>final_data_2 <span class="ot">&lt;-</span> final_data_2[, <span class="sc">-</span><span class="dv">2</span>]</span>
<span id="cb85-5"><a href="#cb85-5"></a>final_data_2 <span class="ot">&lt;-</span> <span class="fu">merge</span>(final_data_2, death, <span class="at">by =</span> <span class="st">"ID"</span>)</span>
<span id="cb85-6"><a href="#cb85-6"></a>final_data_2<span class="sc">$</span>Death <span class="ot">&lt;-</span> <span class="fu">factor</span>(final_data_2<span class="sc">$</span>Death, <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Alive"</span>, <span class="st">"Dead"</span>))</span>
<span id="cb85-7"><a href="#cb85-7"></a></span>
<span id="cb85-8"><a href="#cb85-8"></a>final_data_2_heatmap <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(final_data_2[, <span class="sc">-</span><span class="fu">which</span>(<span class="fu">names</span>(final_data_2) <span class="sc">==</span> <span class="st">"Death"</span>)])</span>
<span id="cb85-9"><a href="#cb85-9"></a>final_data_2_heatmap <span class="ot">&lt;-</span> final_data_2_heatmap[, <span class="sc">-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)]</span>
<span id="cb85-10"><a href="#cb85-10"></a></span>
<span id="cb85-11"><a href="#cb85-11"></a>non_numeric_data <span class="ot">&lt;-</span> final_data_2[, <span class="sc">!</span><span class="fu">sapply</span>(final_data_2, is.numeric)]</span>
<span id="cb85-12"><a href="#cb85-12"></a></span>
<span id="cb85-13"><a href="#cb85-13"></a>orange_palette <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">9</span>, <span class="st">"Oranges"</span>)</span>
<span id="cb85-14"><a href="#cb85-14"></a>group_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Alive"</span> <span class="ot">=</span> <span class="st">"blue"</span>, <span class="st">"Dead"</span> <span class="ot">=</span> <span class="st">"red"</span>)</span>
<span id="cb85-15"><a href="#cb85-15"></a></span>
<span id="cb85-16"><a href="#cb85-16"></a><span class="fu">pheatmap</span>(<span class="fu">t</span>(final_data_2_heatmap),</span>
<span id="cb85-17"><a href="#cb85-17"></a>         <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb85-18"><a href="#cb85-18"></a>         <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>,</span>
<span id="cb85-19"><a href="#cb85-19"></a>         <span class="at">show_rownames =</span> <span class="cn">TRUE</span>, </span>
<span id="cb85-20"><a href="#cb85-20"></a>         <span class="at">show_colnames =</span> <span class="cn">FALSE</span>, </span>
<span id="cb85-21"><a href="#cb85-21"></a>         <span class="at">color =</span> orange_palette,  </span>
<span id="cb85-22"><a href="#cb85-22"></a>         <span class="at">annotation_col =</span> non_numeric_data[, <span class="st">"Death"</span>, <span class="at">drop =</span> <span class="cn">FALSE</span>],  </span>
<span id="cb85-23"><a href="#cb85-23"></a>        <span class="at">annotation_colors =</span> <span class="fu">list</span>(<span class="at">Dead =</span> group_colors),</span>
<span id="cb85-24"><a href="#cb85-24"></a>         <span class="at">legend =</span> <span class="cn">TRUE</span>,</span>
<span id="cb85-25"><a href="#cb85-25"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Hands_on_I_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We are not sure why the colors don’t show in this graph. I apologize because I didn’t have time to finish the second grap.</p>
</section>
</section>
<section id="session-info" class="level1 unnumbered">
<h1 class="unnumbered">session info</h1>
<p>R version 4.4.3 (2025-02-28) Platform: x86_64-pc-linux-gnu Running under: Ubuntu 24.04.2 LTS</p>
<p>Matrix products: default BLAS: /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.12.0 LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.12.0</p>
<p>locale: [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C<br>
[3] LC_TIME=en_US.UTF-8 LC_COLLATE=en_US.UTF-8<br>
[5] LC_MONETARY=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8<br>
[7] LC_PAPER=en_US.UTF-8 LC_NAME=C<br>
[9] LC_ADDRESS=C LC_TELEPHONE=C<br>
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C</p>
<p>time zone: Europe/Madrid tzcode source: system (glibc)</p>
<p>attached base packages: [1] grid stats graphics grDevices utils datasets methods<br>
[8] base</p>
<p>other attached packages: [1] RColorBrewer_1.1-3 circlize_0.4.16 tidyr_1.3.1<br>
[4] kableExtra_1.4.0 gridExtra_2.3 ComplexHeatmap_2.22.0 [7] readxl_1.4.5 pheatmap_1.0.12 ggfortify_0.4.17<br>
[10] interactions_1.2.0 corrplot_0.95 GGally_2.2.1<br>
[13] ggridges_0.5.6 ggplot2_3.5.1 mice_3.17.0<br>
[16] fauxnaif_0.7.1 dplyr_1.1.4</p>
<p>loaded via a namespace (and not attached): [1] Rdpack_2.6.3 rlang_1.1.5 magrittr_2.0.3<br>
[4] clue_0.3-66 GetoptLong_1.0.5 furrr_0.3.1<br>
[7] matrixStats_1.5.0 compiler_4.4.3 png_0.1-8<br>
[10] systemfonts_1.2.1 vctrs_0.6.5 stringr_1.5.1<br>
[13] pkgconfig_2.0.3 shape_1.4.6.1 crayon_1.5.3<br>
[16] fastmap_1.2.0 backports_1.5.0 labeling_0.4.3<br>
[19] utf8_1.2.4 pander_0.6.6 rmarkdown_2.29<br>
[22] nloptr_2.2.1 purrr_1.0.4 xfun_0.51<br>
[25] glmnet_4.1-8 jomo_2.7-6 jsonlite_1.9.1<br>
[28] pan_1.9 broom_1.0.7 parallel_4.4.3<br>
[31] cluster_2.1.8 R6_2.6.1 stringi_1.8.4<br>
[34] parallelly_1.42.0 boot_1.3-31 rpart_4.1.24<br>
[37] cellranger_1.1.0 Rcpp_1.0.14 iterators_1.0.14<br>
[40] knitr_1.50 IRanges_2.40.1 Matrix_1.7-2<br>
[43] splines_4.4.3 nnet_7.3-20 tidyselect_1.2.1<br>
[46] rstudioapi_0.17.1 yaml_2.3.10 doParallel_1.0.17<br>
[49] codetools_0.2-20 listenv_0.9.1 lattice_0.22-5<br>
[52] tibble_3.2.1 plyr_1.8.9 withr_3.0.2<br>
[55] evaluate_1.0.3 future_1.34.0 survival_3.8-3<br>
[58] ggstats_0.9.0 isoband_0.2.7 xml2_1.3.8<br>
[61] jtools_2.3.0 pillar_1.10.1 foreach_1.5.2<br>
[64] stats4_4.4.3 reformulas_0.4.0 generics_0.1.3<br>
[67] S4Vectors_0.44.0 munsell_0.5.1 scales_1.3.0<br>
[70] minqa_1.2.8 globals_0.16.3 glue_1.8.0<br>
[73] tools_4.4.3 lme4_1.1-36 forcats_1.0.0<br>
[76] rbibutils_2.3 colorspace_2.1-1 nlme_3.1-167<br>
[79] cli_3.6.4 viridisLite_0.4.2 svglite_2.1.3<br>
[82] gtable_0.3.6 broom.mixed_0.2.9.6 digest_0.6.37<br>
[85] BiocGenerics_0.52.0 rjson_0.2.23 htmlwidgets_1.6.4<br>
[88] farver_2.1.2 htmltools_0.5.8.1 lifecycle_1.0.4<br>
[91] GlobalOptions_0.1.2 mitml_0.4-5 MASS_7.3-65</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>